// @nekosu/maa-tasker/dist/index.mjs
import { declExpr, makeParser } from "@nekosu/simple-parser";

//#region src/props.ts
const TaskBaseProps = [
	"__baseTaskResolved",
	"baseTask",
	"algorithm",
	"action",
	"sub",
	"subErrorIgnored",
	"next",
	"maxTimes",
	"exceededNext",
	"onErrorNext",
	"preDelay",
	"postDelay",
	"roi",
	"cache",
	"rectMove",
	"reduceOtherTimes",
	"specificRect",
	"specialParams",
	"highResolutionSwipeFix"
];
const TaskExprProps = [
	"sub",
	"next",
	"exceededNext",
	"onErrorNext",
	"reduceOtherTimes"
];
const TaskExprVirts = [
	"sub",
	"next",
	"exceeded_next",
	"on_error_next",
	"reduce_other_times"
];
const TaskExprPropsVirtsMap = {
	sub: "sub",
	next: "next",
	exceededNext: "exceeded_next",
	onErrorNext: "on_error_next",
	reduceOtherTimes: "reduce_other_times"
};
function shouldStrip(prop) {
	return [
		"next",
		"exceeded_next",
		"on_error_next"
	].includes(prop);
}

//#endregion
//#region src/expr/build.ts
function buildExprImpl(ast) {
	switch (ast.type) {
		case "task": return ast.task;
		case "brace": return `(${buildExprImpl(ast.list)})`;
		case "#": return `#${ast.virt}`;
		case "@": return ast.list.map((expr) => buildExprImpl(expr)).join("@") + (ast.virt ? `#${ast.virt}` : "");
		case "*": return `${buildExprImpl(ast.list)} * ${ast.count}`;
		case "+": return `${buildExprImpl(ast.left)} + ${buildExprImpl(ast.right)}`;
		case "^": return `${buildExprImpl(ast.left)} ^ ${buildExprImpl(ast.right)}`;
	}
}

//#endregion
//#region src/expr/parser.ts
let parser = null;
function buildParser() {
	return makeParser([
		["virt", /(?:none|self|back|next|sub|on_error_next|exceeded_next|reduce_other_times)(?![a-zA-Z0-9_-])/],
		["number", /\d+/],
		["task", /[a-zA-Z0-9_-]+/],
		["sharp", /#/],
		["at", /@/],
		["multi", /\*/],
		["plus", /\+/],
		["diff", /\^/],
		["leftBrace", /\(/],
		["rightBrace", /\)/]
	], (curr, getBack) => {
		if (curr === "%virt") return getBack(0) !== "%sharp";
		else if (curr === "%number") return getBack(0) !== "%multi";
		return false;
	}, declExpr(), /[ \t\n]+/, (rule) => {
		return rule.entry("taskList1").do().for("taskVirt").when("%sharp", "%virt").do(([, virt]) => ({
			type: "#",
			virt: virt.value,
			range: virt.range
		})).for("atTaskList").when("taskList4").withloop().when("%at", "taskList4").do(([, task]) => task).do(([task, tasks]) => [task, ...tasks]).for("taskList4").when("%task").do(([task]) => ({
			type: "task",
			task: task.value,
			range: task.range
		})).when("%leftBrace", "taskList1", "%rightBrace").do(([, list]) => ({
			type: "brace",
			list
		})).for("taskList3").sameas("taskList4").when("taskList4", "taskVirt").do(([list, virt]) => ({
			type: "@",
			list: [list],
			virt: virt.virt
		})).sameas("taskVirt").when("atTaskList", "taskVirt").do(([list, virt]) => ({
			type: "@",
			list,
			virt: virt.virt
		})).when("atTaskList").do(([list]) => ({
			type: "@",
			list
		})).for("taskList2").sameas("taskList3").when("taskList3", "%multi", "%number").do(([list, , count]) => ({
			type: "*",
			list,
			count: parseInt(count.value),
			range: count.range
		})).for("taskList1").sameas("taskList2").when("taskList2", "%plus", "taskList1").do(([left, , right]) => ({
			type: "+",
			left,
			right
		})).when("taskList2", "%diff", "taskList1").do(([left, , right]) => ({
			type: "^",
			left,
			right
		}));
	});
}
function getParser() {
	if (!parser) parser = buildParser();
	return parser;
}

//#endregion
//#region src/expr/index.ts
function parseExpr(expr) {
	try {
		return getParser().parse(expr);
	} catch (err) {
		throw `${err}`;
	}
}
function buildExpr(ast) {
	return buildExprImpl(ast);
}

//#endregion
//#region src/eval/utils.ts
function isTaskResolved(task) {
	return !!task.__baseTaskResolved;
}
function isTaskNotResolved(task) {
	return !task.__baseTaskResolved;
}
function removeDuplicated(input, keepLast = false) {
	const inputSet = new Set(input);
	if (keepLast) input = input.toReversed();
	input = input.filter((task) => {
		if (inputSet.has(task)) {
			inputSet.delete(task);
			return true;
		} else return false;
	});
	if (keepLast) input.reverse();
	return input;
}
function removeDuplicatedPrefix(segs) {
	const last = segs.pop();
	return [...removeDuplicated(segs, true), last].join("@");
}
function mergeTask(base, inherit, mode) {
	if (mode === "@" && inherit.task.baseTask) return inherit;
	const result = {
		self: inherit.self,
		task: {},
		trace: {}
	};
	if (inherit.task.algorithm && (base.task.algorithm ?? "MatchTemplate") !== inherit.task.algorithm) {
		for (const key of TaskBaseProps) if (inherit.task[key] !== void 0) {
			result.task[key] = inherit.task[key];
			result.trace[key] = inherit.self;
		} else if (base.task[key] !== void 0) {
			result.task[key] = base.task[key];
			result.trace[key] = base.self;
		}
	} else {
		result.task = { ...base.task };
		result.trace = { ...base.trace };
		for (const [key, val] of Object.entries(inherit.task)) {
			result.task[key] = val;
			result.trace[key] = inherit.self;
		}
		if (!inherit.task.template) {
			delete result.task.template;
			delete result.trace.template;
		}
	}
	if (mode === "baseTask") {
		delete result.task.baseTask;
		delete result.trace.baseTask;
	}
	return result;
}
function mergeMultiPathTasks(tasks) {
	tasks = JSON.parse(JSON.stringify(tasks));
	if (tasks.length === 1) return tasks[0];
	const last = tasks.pop();
	if (last.task.baseTask) {
		if (last.task.baseTask === "#none") {
			delete last.task.baseTask;
			delete last.trace.baseTask;
		}
		return last;
	}
	const prev = mergeMultiPathTasks(tasks);
	prev.self = last.self;
	for (const [key, val] of Object.entries(last.task)) {
		prev.task[key] = val;
		prev.trace[key] = last.self;
	}
	return prev;
}
function applyParentToTask(task, parent) {
	if (!task) return null;
	const clonedTask = JSON.parse(JSON.stringify(task));
	if (!parent) return clonedTask;
	for (const key of TaskExprProps) {
		const prev = clonedTask.task[key];
		if (prev) {
			const newPrev = [];
			for (const expr of prev) newPrev.push(parent.join("@") + expr);
			clonedTask.task[key] = newPrev;
			clonedTask.trace[key] = clonedTask.self;
		}
	}
	return clonedTask;
}

//#endregion
//#region src/eval/context.ts
var MaaErrorDelegate = class {
	taskLoopDetected(tasks) {
		console.error(`task loop detected ${tasks.join(" -> ")}`);
	}
	exprPropLoopDetected(exprs) {
		console.error(`expr loop detected ${exprs.join(" -> ")}`);
	}
	cannotFindTask(task, prefix) {
		console.error(`cannot find task ${task} with parent ${prefix}`);
	}
	warnCannotFindBaseTask(task) {
		console.warn(`cannot find base task ${task}`);
	}
	parseExprError(expr, err) {
		console.error(`parse expr ${expr} failed with error ${err}`);
	}
	exprTooLarge(count) {
		console.error(`expr expand too large ${count}`);
	}
};
var MaaEvalDelegate = class {
	error;
	constructor(error) {
		this.error = error;
	}
	query(_task) {
		return [];
	}
};
var MaaEvalContext = class {
	impl;
	constructor(delegate) {
		this.impl = new MaaEvalContextImpl(delegate);
	}
	evalTask(task) {
		return this.impl.evalTask(task, {
			taskChain: [],
			exprGetPropChain: []
		});
	}
	evalExpr(expr, self, strip = true) {
		return this.impl.evalExpr(expr, self, strip, {
			taskChain: [],
			exprGetPropChain: []
		});
	}
	cleanCache() {
		this.impl.cache = {};
	}
};
var MaaEvalContextImpl = class {
	delegate;
	cache = {};
	constructor(delegate) {
		this.delegate = delegate;
	}
	evalTask(task, context) {
		if (typeof task === "string") task = task.split("@");
		else task = [...task];
		return this.evalTaskImpl(removeDuplicatedPrefix(task), [], context);
	}
	evalExpr(expr, self, strip, context) {
		if (typeof expr === "string") try {
			expr = parseExpr(expr);
		} catch (err) {
			this.delegate.error.parseExprError(expr, `${err}`);
			return null;
		}
		const ast = expr;
		let result;
		switch (ast.type) {
			case "task":
				result = [ast.task];
				break;
			case "brace":
				result = this.evalExpr(ast.list, self, false, context);
				break;
			case "#":
				switch (ast.virt) {
					case "none":
						result = [];
						break;
					case "self":
						result = [self];
						break;
					case "back":
						result = [];
						break;
					case "next":
					case "sub":
					case "exceeded_next":
					case "on_error_next":
					case "reduce_other_times":
						result = [];
						break;
				}
				break;
			case "@": {
				const stages = [];
				let count = 1;
				for (const list of ast.list) {
					const stage = this.evalExpr(list, self, false, context);
					if (!stage) return null;
					stages.push(stage);
					count *= stage.length;
				}
				if (count > 1e5) {
					this.delegate.error.exprTooLarge(count);
					return null;
				}
				while (stages.length > 1) {
					const first = stages.shift();
					const second = stages.shift();
					const output = [];
					for (const f of first) for (const s of second) output.push(`${f}@${s}`);
					stages.unshift(output);
				}
				const expand = stages.shift().map((task) => removeDuplicatedPrefix(task.split("@")));
				switch (ast.virt) {
					case void 0:
						result = expand;
						break;
					case "none":
						result = [];
						break;
					case "self":
						result = expand.map(() => self) ?? null;
						break;
					case "back":
						result = expand;
						break;
					case "next":
					case "sub":
					case "exceeded_next":
					case "on_error_next":
					case "reduce_other_times":
						result = [];
						for (const p of expand) {
							const res = this.getNextList(p, ast.virt, self, context);
							if (!res) return null;
							result.push(...res);
						}
						break;
				}
				break;
			}
			case "*": {
				const list = this.evalExpr(ast.list, self, false, context);
				if (!list) return null;
				result = Array.from({ length: ast.count }, () => [...list]).flat();
				break;
			}
			case "+": {
				const left = this.evalExpr(ast.left, self, false, context);
				const right = this.evalExpr(ast.right, self, false, context);
				if (!left || !right) return null;
				result = [...left, ...right];
				break;
			}
			case "^": {
				const left = this.evalExpr(ast.left, self, false, context);
				const right = this.evalExpr(ast.right, self, false, context);
				if (!left || !right) return null;
				const rightSet = new Set(right);
				result = left.filter((task) => !rightSet.has(task));
				break;
			}
		}
		if (!result) return null;
		if (strip) result = removeDuplicated(result);
		return result;
	}
	evalTaskImpl(task, parent, context) {
		const key = `${parent.join("@")}:${task}`;
		const name = [...parent, task].join("@");
		if (this.cache[name]) return this.cache[name];
		if (context.taskChain.indexOf(key) !== -1) {
			context.taskChain.push(key);
			this.delegate.error.taskLoopDetected(context.taskChain);
			return null;
		}
		context.taskChain.push(key);
		const infos = this.delegate.query(task).map(([obj, anchor]) => {
			const selfAnchor = {
				task: name,
				anchor
			};
			return {
				self: selfAnchor,
				task: obj,
				trace: Object.fromEntries(Object.keys(obj).map((key) => [key, selfAnchor]))
			};
		});
		const segs = task.split("@");
		if (infos.length === 0) {
			if (segs.length === 1) {
				this.delegate.error.cannotFindTask(task, parent);
				context.taskChain.pop();
				return null;
			}
			const seg = segs.shift();
			const result = this.evalTaskImpl(segs.join("@"), [...parent, seg], context);
			context.taskChain.pop();
			return result;
		} else {
			const info = mergeMultiPathTasks(infos);
			if (info.task.baseTask || segs.length === 1) {
				const result = applyParentToTask(this.resolveBaseTask(info, context), parent);
				if (result) this.cache[name] = result;
				context.taskChain.pop();
				return result;
			} else {
				const seg = segs.shift();
				const base = this.evalTask(segs, context);
				if (!base) this.delegate.error.warnCannotFindBaseTask(segs.join("@"));
				let result;
				if (base) {
					this.cache[segs.join("@")] = base;
					const baseWithSeg = applyParentToTask(base, [seg]);
					if (!baseWithSeg) {
						context.taskChain.pop();
						return null;
					}
					result = applyParentToTask(mergeTask(baseWithSeg, info, "@"), parent);
				} else result = applyParentToTask(info, parent);
				if (result) this.cache[name] = result;
				context.taskChain.pop();
				return result;
			}
		}
	}
	resolveBaseTask(task, context) {
		if (isTaskResolved(task.task)) return task;
		else if (isTaskNotResolved(task.task)) {
			if (!task.task.baseTask) return {
				self: task.self,
				task: {
					...task.task,
					__baseTaskResolved: true
				},
				trace: task.trace
			};
			const base = this.evalTask(task.task.baseTask, context);
			if (!base) return null;
			return this.resolveBaseTask(mergeTask(base, task, "baseTask"), context);
		}
		return null;
	}
	getNextList(task, prop, self, context) {
		const key = `${task}.${prop}`;
		if (context.exprGetPropChain.indexOf(key) !== -1) {
			context.exprGetPropChain.push(key);
			this.delegate.error.exprPropLoopDetected(context.exprGetPropChain);
			return null;
		}
		context.exprGetPropChain.push(key);
		const obj = this.evalTask(task, context);
		if (!obj) {
			context.exprGetPropChain.pop();
			return null;
		}
		let rawResult;
		switch (prop) {
			case "next":
				rawResult = obj.task.next ?? [];
				break;
			case "sub":
				rawResult = obj.task.sub ?? [];
				break;
			case "exceeded_next":
				rawResult = obj.task.exceededNext ?? [];
				break;
			case "on_error_next":
				rawResult = obj.task.onErrorNext ?? [];
				break;
			case "reduce_other_times":
				rawResult = obj.task.reduceOtherTimes ?? [];
				break;
		}
		const result = [];
		for (const expr of rawResult) {
			const ast = parseExpr(expr);
			if (!ast) {
				context.exprGetPropChain.pop();
				return null;
			}
			const res = this.evalExpr(ast, self, shouldStrip(prop), context);
			if (!res) {
				context.exprGetPropChain.pop();
				return null;
			}
			result.push(...res);
		}
		context.exprGetPropChain.pop();
		return result;
	}
};

//#endregion
export { MaaErrorDelegate, MaaEvalContext, MaaEvalDelegate, TaskBaseProps, TaskExprProps, TaskExprPropsVirtsMap, TaskExprVirts, buildExpr, parseExpr, shouldStrip };
