// @nekosu/maa-pipeline-manager/dist/logic/index.mjs
import { t } from "@nekosu/maa-locale";

//#region src/logic/runtime/controller.ts
function buildControllerRuntime(data, config) {
	if (config.controller === "$fixed") {
		if (!config.vscFixed) return "No vscFixed for controller";
		if (!config.vscFixed.image) return "No vscFixed image for controller";
		return {
			name: "$fixed",
			type: "vscFixed",
			args: [config.vscFixed.image],
			display_raw: true
		};
	}
	const ctrlInfo = data.controller?.find((x) => x.name === config.controller);
	if (!ctrlInfo) return t("maa.pi.error.cannot-find-controller", config.controller ?? "<unknown>");
	const fixNum = (v, dic) => {
		if (typeof v === "number") return `${v}`;
		else if (dic && typeof v === "string" && v in dic) return dic[v];
		else return v;
	};
	const baseOption = {
		name: ctrlInfo.name,
		display_short_side: ctrlInfo.display_short_side,
		display_long_side: ctrlInfo.display_long_side,
		display_raw: ctrlInfo.display_raw,
		permission_required: ctrlInfo.permission_required,
		attach_resource_path: ctrlInfo.attach_resource_path?.map((x) => x.replaceAll("{PROJECT_DIR}", ".")),
		option: ctrlInfo.option
	};
	if (ctrlInfo.type === "Adb") {
		if (!config.adb) return t("maa.pi.error.cannot-find-adb-for-controller", config.controller ?? "<unknown>");
		return {
			type: "adb",
			args: [
				config.adb.adb_path,
				config.adb.address,
				config.adb.screencap,
				config.adb.input,
				JSON.stringify(config.adb.config)
			],
			...baseOption
		};
	} else if (ctrlInfo.type === "Win32") {
		if (!config.win32) return t("maa.pi.error.cannot-find-win32-for-controller", config.controller ?? "<unknown>");
		if (!config.win32.hwnd) return t("maa.pi.error.cannot-find-hwnd-for-controller", config.controller ?? "<unknown>");
		return {
			type: "win32",
			args: [
				config.win32.hwnd,
				fixNum(ctrlInfo.win32?.screencap, maa.Win32ScreencapMethod) ?? maa.Win32ScreencapMethod.FramePool,
				fixNum(ctrlInfo.win32?.mouse, maa.Win32InputMethod) ?? maa.Win32InputMethod.SendMessageWithCursorPos,
				fixNum(ctrlInfo.win32?.keyboard, maa.Win32InputMethod) ?? maa.Win32InputMethod.SendMessage
			],
			...baseOption
		};
	} else if (ctrlInfo.type === "PlayCover") {
		if (!config.playcover) return t("maa.pi.error.cannot-find-playcover-for-controller", config.controller ?? "<unknown>");
		if (!config.playcover?.address) return t("maa.pi.error.cannot-find-address-for-controller", config.controller ?? "<unknown>");
		return {
			type: "playcover",
			args: [config.playcover.address, "maa.playcover"],
			...baseOption
		};
	} else if (ctrlInfo.type === "Gamepad") {
		if (!config.gamepad) return t("maa.pi.error.cannot-find-gamepad-for-controller", config.controller ?? "<unknown>");
		if (!config.gamepad.hwnd) return t("maa.pi.error.cannot-find-hwnd-for-controller", config.controller ?? "<unknown>");
		return {
			type: "gamepad",
			args: [
				config.gamepad.hwnd,
				fixNum(ctrlInfo.gamepad?.screencap, maa.Win32ScreencapMethod) ?? maa.Win32ScreencapMethod.FramePool,
				fixNum(ctrlInfo.gamepad?.gamepad_type, maa.GamepadType) ?? maa.GamepadType.Xbox360
			],
			...baseOption
		};
	}
	return `Unknown controller type ${ctrlInfo?.type}`;
}

//#endregion
//#region src/logic/runtime/resource.ts
function buildResourceRuntime(data, config) {
	const resInfo = data.resource?.find((info) => info.name === config.resource);
	if (!resInfo) return t("maa.pi.error.cannot-find-resource", config.resource ?? "");
	const paths = (typeof resInfo.path === "string" ? [resInfo.path] : resInfo.path).map((x) => x.replaceAll("{PROJECT_DIR}", "."));
	return {
		name: resInfo.name,
		paths,
		option: resInfo.option
	};
}

//#endregion
//#region src/logic/runtime/option.ts
function isStringArray(arr) {
	return Array.isArray(arr) && !arr.find((x) => typeof x !== "string");
}
function isStringStringObject(obj) {
	return typeof obj === "object" && obj !== null && !Object.values(obj).find((x) => typeof x !== "string");
}
function resolveOptionConfig(task, option, type) {
	const val = task.option?.[option];
	switch (type) {
		case "select":
		case "switch": return typeof val === "string" ? val : void 0;
		case "checkbox": return isStringArray(val) ? val : void 0;
		case "input": return isStringStringObject(val) ? val : void 0;
	}
}
function resolveSelect(task, option, optMeta) {
	const cfg = resolveOptionConfig(task, option, "select") ?? optMeta.default_case ?? optMeta.cases?.[0].name;
	if (!cfg) return null;
	return optMeta.cases?.find((x) => x.name === cfg) ?? null;
}
function resolveCheckbox(task, option, optMeta) {
	const cfg = resolveOptionConfig(task, option, "checkbox") ?? [];
	return optMeta.cases?.filter((x) => cfg.includes(x.name)) ?? null;
}
function buildOption(data, task, ctrlRt, resRt) {
	const taskInfo = data.task?.find((x) => x.name === task.name);
	if (!taskInfo) return t("maa.pi.error.cannot-find-task", task.name);
	const pending = [];
	for (const opt of data.global_option ?? []) pending.push({
		name: opt,
		from: "global",
		origin: ""
	});
	for (const opt of ctrlRt.option ?? []) pending.push({
		name: opt,
		from: "controller",
		origin: ctrlRt.name
	});
	for (const opt of resRt.option ?? []) pending.push({
		name: opt,
		from: "resource",
		origin: resRt.name
	});
	for (const opt of taskInfo.option ?? []) pending.push({
		name: opt,
		from: "task",
		origin: taskInfo.name
	});
	const resolved = [];
	const resolvedOption = /* @__PURE__ */ new Set();
	while (pending.length > 0) {
		const opt = pending.shift();
		if (resolvedOption.has(opt.name)) continue;
		resolved.push(opt);
		resolvedOption.add(opt.name);
		const optMeta = data.option?.[opt.name];
		if (!optMeta) return t("maa.pi.error.cannot-find-option-from", opt.name, opt.from, opt.origin);
		if (ctrlRt.name !== "$fixed" && optMeta.controller && !optMeta.controller.includes(ctrlRt.name)) continue;
		if (optMeta.resource && !optMeta.resource.includes(resRt.name)) continue;
		if (!optMeta.type || optMeta.type === "select" || optMeta.type === "switch") {
			const caseMeta = resolveSelect(task, opt.name, optMeta);
			if (!caseMeta) return t("maa.pi.error.cannot-resolve-option", opt.name);
			for (const sub of caseMeta.option ?? []) pending.push({
				name: sub,
				from: "option",
				origin: opt.name
			});
		} else if (optMeta.type === "checkbox") {
			const caseMetas = resolveCheckbox(task, opt.name, optMeta);
			if (!caseMetas) return t("maa.pi.error.cannot-resolve-option", opt.name);
			for (const caseMeta of caseMetas) for (const sub of caseMeta.option ?? []) pending.push({
				name: sub,
				from: "option",
				origin: opt.name
			});
		}
	}
	return resolved;
}

//#endregion
//#region src/logic/runtime/task.ts
function buildTask(data, task, ctrlRt, resRt) {
	const overrides = [];
	const options = buildOption(data, task, ctrlRt, resRt);
	if (typeof options === "string") return options;
	for (const opt of options) {
		const optMeta = data.option?.[opt.name];
		if (!optMeta) return t("maa.pi.error.cannot-find-option-from", opt.name, opt.from, opt.origin);
		if (!optMeta.type || optMeta.type === "select" || optMeta.type === "switch") {
			const caseMeta = resolveSelect(task, opt.name, optMeta);
			if (caseMeta?.pipeline_override) overrides.push(caseMeta.pipeline_override);
		} else if (optMeta.type === "checkbox") {
			const caseMetas = resolveCheckbox(task, opt.name, optMeta);
			for (const caseMeta of caseMetas ?? []) if (caseMeta?.pipeline_override) overrides.push(caseMeta.pipeline_override);
		} else if (optMeta.type === "input") {
			const vals = resolveOptionConfig(task, opt.name, "input") ?? {};
			const updateOverride = (v) => {
				if (Array.isArray(v)) return v.map(updateOverride);
				else if (typeof v === "object" && v !== null) {
					const obj = v;
					return Object.fromEntries(Object.entries(obj).map(([key, val]) => {
						return [key, updateOverride(val)];
					}));
				} else if (typeof v === "string") {
					let finalType = void 0;
					let result = v;
					for (const subOpt of optMeta.inputs ?? []) if (result.indexOf(`{${subOpt.name}}`) !== -1) {
						const expectType = subOpt.pipeline_type ?? "string";
						if (finalType && finalType !== expectType) throw "input type mismatch!";
						finalType = expectType;
						result = result.replaceAll(`{${subOpt.name}}`, vals[subOpt.name]);
					}
					switch (finalType) {
						case "string": return result;
						case "int": return parseInt(result);
						case "bool": return result === "true";
					}
					return "";
				} else return v;
			};
			try {
				overrides.push(updateOverride(optMeta.pipeline_override ?? {}));
			} catch (err) {
				return `${err}`;
			}
		}
	}
	return overrides;
}
function buildTaskRuntime(data, config, ctrlRt, resRt) {
	const taskRt = { tasks: [] };
	for (const task of config.task ?? []) {
		const taskInfo = data.task?.find((x) => x.name === task.name);
		if (!taskInfo) return t("maa.pi.error.cannot-find-task", task.name);
		const info = buildTask(data, task, ctrlRt, resRt);
		if (typeof info === "string") return info;
		taskRt.tasks.push({
			name: task.name,
			entry: taskInfo.entry,
			pipeline_override: info
		});
	}
	return taskRt;
}

//#endregion
export { buildControllerRuntime, buildOption, buildResourceRuntime, buildTask, buildTaskRuntime, resolveCheckbox, resolveOptionConfig, resolveSelect };
