// @nekosu/maa-pipeline-manager/dist/index.mjs
import { buildControllerRuntime, buildOption, buildResourceRuntime, buildTask, buildTaskRuntime, resolveCheckbox, resolveOptionConfig, resolveSelect } from "./logic/index.mjs";
import EventEmitter from "node:events";
import * as path from "node:path";
import { parseTree } from "jsonc-parser";
import { MaaErrorDelegate, MaaEvalContext, MaaEvalDelegate, parseExpr } from "@nekosu/maa-tasker";
import * as fs from "node:fs/promises";
import chokidar from "chokidar";
import { t } from "@nekosu/maa-locale";

//#region src/parser/task/keys.ts
const nodeKeys = [
	"next",
	"rate_limit",
	"timeout",
	"on_error",
	"anchor",
	"inverse",
	"enabled",
	"max_hit",
	"pre_delay",
	"post_delay",
	"pre_wait_freezes",
	"post_wait_freezes",
	"repeat",
	"repeat_delay",
	"repeat_wait_freezes",
	"focus",
	"attach",
	"doc",
	"desc",
	"sub_name"
];
const recoKeys = [
	"roi",
	"roi_offset",
	"template",
	"threshold",
	"order_by",
	"index",
	"method",
	"green_mask",
	"count",
	"detector",
	"ratio",
	"lower",
	"upper",
	"connected",
	"expected",
	"replace",
	"only_rec",
	"model",
	"color_filter",
	"labels",
	"all_of",
	"box_index",
	"any_of",
	"custom_recognition",
	"custom_recognition_param"
];
const actKeys = [
	"target",
	"target_offset",
	"contact",
	"pressure",
	"duration",
	"begin",
	"begin_offset",
	"end",
	"end_offset",
	"end_hold",
	"only_hover",
	"swipes",
	"dx",
	"dy",
	"key",
	"input_text",
	"package",
	"exec",
	"args",
	"detach",
	"cmd",
	"shell_timeout",
	"filename",
	"format",
	"quality",
	"custom_action",
	"custom_action_param"
];
const maaNodeKeys = [
	"baseTask",
	"sub",
	"subErrorIgnored",
	"next",
	"maxTimes",
	"exceededNext",
	"onErrorNext",
	"preDelay",
	"postDelay",
	"roi",
	"cache",
	"rectMove",
	"reduceOtherTimes",
	"specificRect",
	"specialParams",
	"highResolutionSwipeFix"
];
const maaRecoKeys = [
	"template",
	"templThreshold",
	"maskRange",
	"colorScales",
	"colorWithClose",
	"pureColor",
	"method",
	"text",
	"ocrReplace",
	"fullMatch",
	"isAscii",
	"withoutDet",
	"useRaw",
	"binThreshold",
	"count",
	"ratio",
	"detector"
];
const maaActKeys = ["inputText"];

//#endregion
//#region src/parser/utils.ts
function parseProp(prop) {
	const pair = parsePropFlex(prop);
	if (!pair) return null;
	const [key, obj, node] = pair;
	if (!obj) return null;
	return [
		key,
		obj,
		node
	];
}
function parsePropFlex(prop) {
	if (prop.type === "property" && prop.children?.length === 1 && isString(prop.children[0])) return [
		prop.children[0].value,
		null,
		prop.children[0]
	];
	if (prop.type !== "property" || !prop.children || prop.children.length !== 2) return null;
	const [key, obj] = prop.children;
	if (!isString(key)) return null;
	return [
		key.value,
		obj,
		key
	];
}
function* parseObject(node) {
	if (!node || node.type !== "object") return;
	for (const prop of node.children ?? []) {
		const pair = parseProp(prop);
		if (pair) yield pair;
	}
}
function* parseObjectFlex(node) {
	if (!node || node.type !== "object") return;
	for (const prop of node.children ?? []) {
		const pair = parsePropFlex(prop);
		if (pair) yield pair;
	}
}
function* parseArray(node) {
	if (!node || node.type !== "array") return;
	for (const obj of node.children ?? []) yield obj;
}
function isString(node) {
	return !!node && node.type === "string" && typeof node.value === "string";
}
function isNumber(node) {
	return !!node && node.type === "number" && typeof node.value === "number";
}
function isBool(node) {
	return !!node && node.type === "boolean" && typeof node.value === "boolean";
}

//#endregion
//#region src/utils/json.ts
function shrinkParent(node) {
	delete node.parent;
	for (const child of node.children ?? []) shrinkParent(child);
}
function parseTreeWithoutParent(content) {
	const node = parseTree(content);
	if (node) shrinkParent(node);
	return node;
}
function buildTree(node) {
	switch (node.type) {
		case "string":
		case "number":
		case "boolean": return node.value ?? null;
		case "object": return Object.fromEntries([...parseObject(node)].map(([key, obj]) => [key, buildTree(obj)]));
		case "array": return [...parseArray(node)].map(buildTree);
		case "property": return null;
		case "null": return null;
	}
	return null;
}

//#endregion
//#region src/utils/types.ts
function joinPath(...segs) {
	return path.join(...segs);
}
function joinImagePath(maa, root, image) {
	return path.join(root, maa ? "template" : "image", image);
}
function normalizeImageFolder(image) {
	let norm = path.normalize(image).replaceAll(path.sep, "/");
	if (norm.endsWith("/")) norm = norm.slice(0, -1);
	return norm;
}
function relativePath(base, target) {
	return path.relative(base, target);
}

//#endregion
//#region src/layer/layer.ts
function specialStringify(value, indent, indentCount) {
	if (Array.isArray(value)) {
		if (value.length === 0) return "[]";
		const result = ["["];
		for (const val of value) result.push(indent.repeat(indentCount) + specialStringify(val, indent, indentCount + 1) + ",");
		result.push(indent.repeat(indentCount - 1) + "]");
		return result.join("\n");
	} else if (typeof value === "object" && value !== null) {
		if (Object.keys(value).length === 0) return "{}";
		const result = ["{"];
		for (const [key, val] of Object.entries(value)) result.push(indent.repeat(indentCount) + JSON.stringify(key) + ": " + specialStringify(val, indent, indentCount + 1) + ",");
		result.push(indent.repeat(indentCount - 1) + "}");
		return result.join("\n");
	} else return JSON.stringify(value);
}
var LayerInfo = class {
	loader;
	maa;
	root;
	parent;
	type;
	tasks;
	images;
	extraDecls;
	extraRefs;
	dirty;
	mergedDeclsCache;
	mergedRefsCache;
	constructor(loader, maa, root, type) {
		this.loader = loader;
		this.maa = maa;
		this.root = root;
		this.type = type;
		this.tasks = {};
		this.images = /* @__PURE__ */ new Set();
		this.extraDecls = [];
		this.extraRefs = [];
		this.dirty = true;
		this.mergedDeclsCache = [];
		this.mergedRefsCache = [];
	}
	reset() {
		this.tasks = {};
		this.images = /* @__PURE__ */ new Set();
		this.extraDecls = [];
		this.extraRefs = [];
		this.dirty = true;
		this.mergedDeclsCache = [];
		this.mergedRefsCache = [];
	}
	mutableTaskInfo(name) {
		this.tasks[name] = this.tasks[name] ?? [];
		return this.tasks[name];
	}
	removeFile(file) {
		const changed = [];
		for (const [task, infos] of Object.entries(this.tasks)) {
			const newInfos = infos.filter((info) => info.file !== file);
			if (infos.length !== newInfos.length) {
				if (newInfos.length === 0) delete this.tasks[task];
				else infos.splice(0, infos.length, ...newInfos);
				changed.push(task);
			}
		}
		this.extraDecls = this.extraDecls.filter((decl) => decl.file !== file);
		this.extraRefs = this.extraRefs.filter((ref) => ref.file !== file);
		this.markDirty();
		return changed;
	}
	markDirty() {
		this.dirty = true;
	}
	get mergedDecls() {
		this.flushMergedDeclsRefs();
		return this.mergedDeclsCache;
	}
	get mergedRefs() {
		this.flushMergedDeclsRefs();
		return this.mergedRefsCache;
	}
	get mergedAllDecls() {
		return (this.parent?.mergedAllDecls ?? []).concat(this.mergedDecls);
	}
	get mergedAllRefs() {
		return (this.parent?.mergedAllRefs ?? []).concat(this.mergedRefs);
	}
	flushMergedDeclsRefs() {
		if (!this.dirty) return;
		this.mergedDeclsCache = [];
		this.mergedRefsCache = [];
		for (const taskInfos of Object.values(this.tasks)) for (const taskInfo of taskInfos) {
			this.mergedDeclsCache.push(...taskInfo.info.decls);
			this.mergedRefsCache.push(...taskInfo.info.refs);
		}
		this.mergedDeclsCache.push(...this.extraDecls);
		this.mergedRefsCache.push(...this.extraRefs);
		this.dirty = false;
	}
	getTaskListNotUnique() {
		return (this.parent?.getTaskList() ?? []).concat(Object.keys(this.tasks).filter((task) => !task.startsWith("$")));
	}
	getTaskList() {
		return [...new Set(this.getTaskListNotUnique())];
	}
	getAnchorList() {
		const anchors = this.parent?.getAnchorList() ?? [];
		const decls = this.mergedDecls.filter((decl) => decl.type === "task.anchor");
		anchors.push(...decls.map((decl) => [decl.anchor, decl]));
		return anchors;
	}
	getImageListNotUnique() {
		return (this.parent?.getImageList() ?? []).concat(...this.images);
	}
	getImageList() {
		return [...new Set(this.getImageListNotUnique())];
	}
	getImageFolders() {
		const result = this.parent?.getImageFolders() ?? /* @__PURE__ */ new Map();
		for (const image of this.images) {
			const rel = path.dirname(image);
			if (result.has(rel)) {
				const arr = result.get(rel);
				if (arr[0] !== this) arr.unshift(this);
			} else result.set(rel, [this]);
		}
		return result;
	}
	maaFindTaskDecl(task) {
		const tasks = this.getTaskList();
		let current = task;
		while (!tasks.includes(current) && current.indexOf("@") !== -1) current = current.replace(/^[^@]+@/, "");
		return current;
	}
	getTask(task, maaTrace = true) {
		const tasks = this.parent?.getTask(task) ?? [];
		const infos = {
			layer: this,
			infos: [...this.tasks[task] ?? []]
		};
		tasks.unshift(infos);
		if (this.maa && maaTrace) {
			let current = task;
			while (current.indexOf("@") !== -1) {
				const next = current.replace(/^[^@]+@/, "");
				infos.infos.push(...this.tasks[next] ?? []);
				current = next;
			}
		}
		return tasks.filter((x) => x.infos.length > 0);
	}
	evalTask(task) {
		const upper = this.parent?.evalTask(task);
		const result = upper ?? {};
		const info = this.tasks[task]?.[0];
		if (info) {
			const parts = info.info.parts;
			if (!upper) {
				const reco = "$" + (parts.recoType?.value ?? "DirectHit");
				const act = "$" + (parts.actType?.value ?? "DoNothing");
				Object.assign(result, this.tasks["$Default"]?.[0].obj ?? {});
				Object.assign(result, this.tasks[reco]?.[0].obj ?? {});
				Object.assign(result, this.tasks[act]?.[0].obj ?? {});
			}
			let recoChanged = false;
			let actChanged = false;
			if (parts.recoType) {
				const oldReco = result.recognition ?? "DirectHit";
				recoChanged = parts.recoType.value !== oldReco;
				result["recognition"] = parts.recoType.value;
			}
			if (parts.actType) {
				const oldAct = result.action ?? "DoNothing";
				actChanged = parts.actType.value !== oldAct;
				result["action"] = parts.actType.value;
			}
			if (recoChanged) for (const key of recoKeys) delete result[key];
			if (actChanged) for (const key of actKeys) delete result[key];
			for (const [key, obj] of [
				...parts.base,
				...parts.reco,
				...parts.act,
				...parts.unknown
			]) result[key] = buildTree(obj);
		}
		return result;
	}
	getImage(image) {
		const layers = this.parent?.getImage(image) ?? [];
		if (this.images.has(image)) layers.unshift([
			this,
			joinImagePath(this.maa, this.root, image),
			image
		]);
		if (this.maa) {
			const suffix = "/" + image;
			for (const file of this.images) if (file.endsWith(suffix)) layers.unshift([
				this,
				joinImagePath(this.maa, this.root, file),
				file
			]);
		}
		return layers;
	}
	getTaskBriefInfo(task) {
		const result = {};
		for (const { infos } of this.getTask(task)) for (const info of infos) {
			if (!result.reco && info.info.parts.recoType) result.reco = info.info.parts.recoType.value;
			else if (!result.act && info.info.parts.actType) result.act = info.info.parts.actType.value;
			if (result.reco && result.act) return result;
		}
		return result;
	}
	getTaskDoc(task) {
		return this.mergedAllDecls.filter((decl) => decl.type === "task.doc").filter((decl) => decl.task === task).map((decl) => decl.doc).join(" ");
	}
	toggleMode(mode, info, indent = "    ") {
		const parts = info.info.parts;
		const data = {};
		if (mode === 1) {
			if (parts.recoType) data.recognition = parts.recoType.value;
			for (const [key, obj] of parts.reco) data[key] = buildTree(obj);
			if (parts.actType) data.action = parts.actType.value;
			for (const [key, obj] of parts.act) data[key] = buildTree(obj);
		} else if (mode === 2) {
			if (parts.recoType || parts.reco.length > 0) {
				data.recognition = {};
				if (parts.recoType) data.recognition.type = parts.recoType.value;
				if (parts.reco.length > 0) {
					data.recognition.param = {};
					for (const [key, obj] of parts.reco) data.recognition.param[key] = buildTree(obj);
				}
			}
			if (parts.actType || parts.act.length > 0) {
				data.action = {};
				if (parts.actType) data.action.type = parts.actType.value;
				if (parts.act.length > 0) {
					data.action.param = {};
					for (const [key, obj] of parts.act) data.action.param[key] = buildTree(obj);
				}
			}
		}
		for (const [key, obj] of parts.base) data[key] = buildTree(obj);
		for (const [key, obj] of parts.unknown) data[key] = buildTree(obj);
		return JSON.stringify(info.prop.value) + ": " + specialStringify(data, indent, 2);
	}
};

//#endregion
//#region src/parser/task/anchor.ts
function parseSingle$3(node, info, ctx) {
	if (isString(node)) info.decls.push({
		file: ctx.file,
		location: node,
		type: "task.anchor",
		anchor: node.value,
		task: ctx.taskName,
		belong: ctx.taskName
	});
}
function parseAnchor(node, info, ctx) {
	if (isString(node)) parseSingle$3(node, info, ctx);
	else if (node.type === "array") for (const obj of parseArray(node)) parseSingle$3(obj, info, ctx);
	else for (const [key, obj, prop] of parseObjectFlex(node)) if (obj && isString(obj)) {
		info.decls.push({
			file: ctx.file,
			location: prop,
			type: "task.anchor",
			anchor: key,
			task: obj.value,
			belong: ctx.taskName
		});
		info.refs.push({
			file: ctx.file,
			location: obj,
			type: "task.anchor",
			target: obj.value
		});
	} else info.decls.push({
		file: ctx.file,
		location: prop,
		type: "task.anchor",
		anchor: key,
		task: "",
		belong: ctx.taskName
	});
}

//#endregion
//#region src/parser/task/color.ts
function isColor(node) {
	let length = 0;
	for (const obj of parseArray(node)) {
		if (!isNumber(obj)) return false;
		length += 1;
	}
	return length === 3;
}
function parseColorSingle(node, info, ctx, method) {
	const color = [];
	for (const obj of parseArray(node)) if (isNumber(obj)) color.push(obj.value);
	info.refs.push({
		location: node,
		file: ctx.file,
		type: "task.color",
		method,
		color
	});
}
function parseColor(node, info, ctx, method) {
	if (isColor(node)) parseColorSingle(node, info, ctx, method);
	else for (const item of parseArray(node)) if (isColor(item)) parseColorSingle(item, info, ctx, method);
}

//#endregion
//#region src/parser/task/colorFilter.ts
function parseColorFilter(node, info, ctx) {
	if (isString(node)) info.refs.push({
		file: ctx.file,
		location: node,
		type: "task.color_filter",
		target: node.value
	});
}

//#endregion
//#region src/parser/task/focus.ts
function parseFocus(node, info, ctx) {
	for (const [_key, obj] of parseObject(node)) if (isString(obj)) {
		if (obj.value.startsWith("$")) info.refs.push({
			file: ctx.file,
			location: obj,
			type: "task.locale",
			target: obj.value.substring(1)
		});
		else if (obj.value.length > 0) info.refs.push({
			file: ctx.file,
			location: obj,
			type: "task.can_locale",
			target: obj.value
		});
	}
}

//#endregion
//#region src/parser/task/attr.ts
function parseAttr(name, keys) {
	const info = {
		offset: 0,
		attrs: {},
		unknown: []
	};
	let offset = 0;
	while (true) {
		let found = false;
		for (const key of keys) {
			const prefix = `[${key}]`;
			if (name.startsWith(prefix)) {
				info.attrs[key] = true;
				name = name.substring(prefix.length);
				offset += prefix.length;
				found = true;
				break;
			}
		}
		if (found) continue;
		const match = /^\[([^\]]+)\]/.exec(name);
		if (match) {
			info.unknown.push([
				match[1],
				offset,
				match[1].length + 2
			]);
			name = name.substring(match[1].length + 2);
			offset += match[1].length + 2;
			continue;
		}
		break;
	}
	info.offset = offset;
	return [name, info];
}

//#endregion
//#region src/parser/task/target.ts
function parseTarget(node, info, ctx, acceptArray = false) {
	if (isString(node)) {
		const [target, attrs] = parseAttr(node.value, ["Anchor"]);
		info.refs.push({
			file: ctx.file,
			location: node,
			type: "task.target",
			target,
			attrs
		});
	} else if (acceptArray) for (const obj of parseArray(node)) parseTarget(obj, info, ctx);
}

//#endregion
//#region src/parser/task/freeze.ts
function parseFreeze(node, info, ctx) {
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "target":
			parseTarget(obj, info, ctx);
			break;
	}
}

//#endregion
//#region src/parser/task/maa/baseTask.ts
function parseMaaBaseTask(node, info, ctx) {
	if (isString(node)) info.refs.push({
		file: ctx.file,
		location: node,
		type: "task.maa.base_task",
		target: node.value,
		tasks: buildTaskRef(node.value),
		belong: ctx.taskName
	});
}

//#endregion
//#region src/parser/task/maa/expr.ts
function calcSuffix(list) {
	if (list.length === 0) return;
	let current = list[0].task;
	list.shift();
	while (list.length > 0) {
		const next = list.shift();
		current = `${next.task}@${current}`;
		next.taskSuffix = current;
	}
}
function parseMaaExprTask(ast, tasks) {
	switch (ast.type) {
		case "task":
			tasks.push({
				task: ast.task,
				taskSuffix: ast.task,
				offset: ast.range[0],
				length: ast.range[1]
			});
			return tasks[tasks.length - 1];
		case "brace":
			parseMaaExprTask(ast.list, tasks);
			break;
		case "@": {
			let list = [];
			for (const sub of ast.list) {
				const next = parseMaaExprTask(sub, tasks);
				if (next) list.unshift(next);
				else {
					calcSuffix(list);
					list = [];
				}
			}
			calcSuffix(list);
			break;
		}
		case "#": break;
		case "*":
			parseMaaExprTask(ast.list, tasks);
			break;
		case "+":
		case "^":
			parseMaaExprTask(ast.left, tasks);
			parseMaaExprTask(ast.right, tasks);
			break;
	}
}
function parseMaaExpr(node, info, ctx) {
	if (isString(node)) {
		const tasks = [];
		info.refs.push({
			file: ctx.file,
			location: node,
			type: "task.maa.expr",
			target: node.value,
			tasks,
			belong: ctx.taskName
		});
		try {
			parseMaaExprTask(parseExpr(node.value), tasks);
		} catch (_err) {
			return;
		}
	}
}
function parseMaaExprList(node, info, ctx) {
	for (const obj of parseArray(node)) parseMaaExpr(obj, info, ctx);
}

//#endregion
//#region src/parser/task/next.ts
function parseSingle$2(node, info, ctx) {
	if (isString(node)) {
		const [target, attrs] = parseAttr(node.value, ["JumpBack", "Anchor"]);
		info.refs.push({
			file: ctx.file,
			location: node,
			type: "task.next",
			target,
			objMode: false,
			attrs
		});
	} else if (node.type === "object") {
		let loc = null;
		const ref = {
			type: "task.next",
			target: "",
			objMode: true,
			attrs: {
				offset: 0,
				attrs: {},
				unknown: []
			}
		};
		for (const [key, obj] of parseObject(node)) if (key === "name" && isString(obj)) {
			ref.target = obj.value;
			loc = obj;
		} else if (key === "jump_back" && isBool(obj)) ref.attrs.attrs.JumpBack = obj.value;
		else if (key === "anchor" && isBool(obj)) ref.attrs.attrs.Anchor = obj.value;
		if (loc) info.refs.push({
			file: ctx.file,
			location: loc,
			...ref
		});
	}
}
function parseNextList(node, info, ctx, forceArray = false) {
	if (!forceArray && node.type !== "array") parseSingle$2(node, info, ctx);
	else for (const obj of parseArray(node)) parseSingle$2(obj, info, ctx);
}

//#endregion
//#region src/parser/task/roi.ts
function parseRoi(node, info, prev, ctx) {
	if (isString(node)) {
		const [target, attrs] = parseAttr(node.value, ["Anchor"]);
		if (attrs.offset > 0) info.refs.push({
			file: ctx.file,
			location: node,
			type: "task.roi",
			target,
			attrs,
			prev: [...prev],
			task: ctx.taskName,
			prevRef: false
		});
		else {
			const prevRef = !!prev.find((decl) => decl.value === node.value);
			info.refs.push({
				file: ctx.file,
				location: node,
				type: "task.roi",
				target: node.value,
				attrs: {
					offset: 0,
					attrs: {},
					unknown: []
				},
				prev: [...prev],
				task: ctx.taskName,
				prevRef
			});
		}
	}
}

//#endregion
//#region src/parser/task/split.ts
function splitNode(node, maa) {
	const result = {
		node,
		base: [],
		reco: [],
		act: [],
		unknown: []
	};
	if (maa) {
		for (const pair of parseObject(node)) {
			const [key, obj] = pair;
			if (key === "algorithm" && isString(obj)) result.recoType = obj;
			else if (key === "action" && isString(obj)) result.actType = obj;
			else if (maaNodeKeys.includes(key)) result.base.push(pair);
			else if (maaRecoKeys.includes(key)) result.reco.push(pair);
			else if (maaActKeys.includes(key)) result.act.push(pair);
			else result.unknown.push(pair);
		}
		return result;
	}
	for (const pair of parseObject(node)) {
		const [key, obj] = pair;
		if (nodeKeys.includes(key)) result.base.push(pair);
		else if (recoKeys.includes(key)) result.reco.push(pair);
		else if (actKeys.includes(key)) result.act.push(pair);
		else if (key === "recognition") {
			if (isString(obj)) result.recoType = obj;
			else if (obj.type === "object") {
				const type = obj.children?.find((node) => node.children?.[0].value === "type" && isString(node.children?.[1]));
				const param = obj.children?.find((node) => node.children?.[0].value === "param");
				if (type) result.recoType = type.children[1];
				for (const pair of parseObject(param?.children?.[1])) if (recoKeys.includes(pair[0])) result.reco.push(pair);
			}
		} else if (key === "action") {
			if (isString(obj)) result.actType = obj;
			else if (obj.type === "object") {
				const type = obj.children?.find((node) => node.children?.[0].value === "type" && isString(node.children?.[1]));
				const param = obj.children?.find((node) => node.children?.[0].value === "param");
				if (type) result.actType = type.children[1];
				for (const pair of parseObject(param?.children?.[1])) if (actKeys.includes(pair[0])) result.act.push(pair);
			}
		} else result.unknown.push(pair);
	}
	return result;
}

//#endregion
//#region src/parser/task/subName.ts
function parseSubName(node, info, parent, ctx) {
	if (isString(node)) {
		info.decls.push({
			file: ctx.file,
			location: node,
			type: "task.sub_reco",
			name: node.value,
			reco: parent,
			task: ctx.taskName
		});
		return node;
	} else return null;
}

//#endregion
//#region src/parser/task/template.ts
function parseSingle$1(node, info, ctx) {
	if (isString(node)) info.refs.push({
		file: ctx.file,
		location: node,
		type: "task.template",
		target: node.value
	});
}
function parseTemplate(node, info, ctx) {
	if (node.type !== "array") parseSingle$1(node, info, ctx);
	else for (const obj of parseArray(node)) parseSingle$1(obj, info, ctx);
}

//#endregion
//#region src/parser/task/task.ts
function parseMaaBase(props, info, ctx) {
	for (const [prop, obj] of props) switch (prop) {
		case "baseTask":
			parseMaaBaseTask(obj, info, ctx);
			break;
		case "sub":
		case "next":
		case "exceededNext":
		case "onErrorNext":
		case "reduceOtherTimes":
			parseMaaExprList(obj, info, ctx);
			break;
	}
}
function parseBase(props, info, ctx) {
	for (const [key, obj] of props) switch (key) {
		case "next":
		case "on_error":
			parseNextList(obj, info, ctx);
			break;
		case "anchor":
			parseAnchor(obj, info, ctx);
			break;
		case "pre_wait_freezes":
		case "post_wait_freezes":
		case "repeat_wait_freezes":
			parseFreeze(obj, info, ctx);
			break;
		case "focus":
			parseFocus(obj, info, ctx);
			break;
		case "doc":
		case "desc":
			if (isString(obj)) info.decls.push({
				file: ctx.file,
				location: obj,
				type: "task.doc",
				task: ctx.taskName,
				doc: obj.value
			});
			break;
	}
}
function parseMaaReco(props, info, ctx) {
	for (const [prop, obj] of props) switch (prop) {
		case "template":
			parseTemplate(obj, info, ctx);
			break;
	}
}
function parseReco(props, baseProps, info, prev, ctx, parent) {
	let subName = null;
	let colorMatchMethod = "rgb";
	for (const [key, obj] of props) switch (key) {
		case "roi":
			parseRoi(obj, info, prev, ctx);
			break;
		case "template":
			parseTemplate(obj, info, ctx);
			break;
		case "color_filter":
			parseColorFilter(obj, info, ctx);
			break;
		case "all_of":
		case "any_of":
			for (const sub of parseArray(obj)) if (isString(sub)) info.refs.push({
				file: ctx.file,
				location: sub,
				type: "task.reco",
				target: sub.value
			});
			else {
				const subInfo = splitNode(sub, false);
				parseReco(subInfo.reco, subInfo.base, info, prev, ctx, sub);
			}
			break;
		case "method":
			if (isNumber(obj)) switch (obj.value) {
				case 4:
					colorMatchMethod = "rgb";
					break;
				case 40:
					colorMatchMethod = "hsv";
					break;
				default: colorMatchMethod = null;
			}
			break;
	}
	for (const [key, obj] of baseProps) switch (key) {
		case "sub_name":
			if (parent) subName = parseSubName(obj, info, parent, ctx);
			break;
	}
	if (subName) prev.push(subName);
	if (colorMatchMethod) for (const [key, obj] of props) switch (key) {
		case "upper":
		case "lower":
			parseColor(obj, info, ctx, colorMatchMethod);
			break;
	}
}
function parseAct(props, info, ctx) {
	for (const [key, obj] of props) switch (key) {
		case "target":
		case "begin":
			parseTarget(obj, info, ctx);
			break;
		case "end":
			parseTarget(obj, info, ctx, true);
			break;
	}
}
function parseUnknown(props, info, ctx) {
	for (const [key, _obj, prop] of props) if (key.startsWith("$__mpe")) info.decls.push({
		file: ctx.file,
		location: prop,
		type: "task.mpe_config"
	});
}
function buildTaskRef(task) {
	let offset = 0;
	const tasks = task.split("@").map((task) => {
		const result = {
			task,
			taskSuffix: task,
			offset,
			length: task.length
		};
		offset += task.length + 1;
		return result;
	});
	let suffix = tasks[tasks.length - 1].task;
	for (let idx = tasks.length - 2; idx >= 0; idx--) {
		suffix = `${tasks[idx].task}@${suffix}`;
		tasks[idx].taskSuffix = suffix;
	}
	return tasks;
}
function parseTask(node, ctx) {
	const parts = splitNode(node, ctx.maa);
	const info = {
		parts,
		decls: [],
		refs: []
	};
	info.decls.push({
		file: ctx.file,
		location: ctx.task,
		type: "task.decl",
		task: ctx.taskName,
		tasks: buildTaskRef(ctx.taskName)
	});
	if (ctx.maa) {
		parseMaaBase(info.parts.base, info, ctx);
		parseMaaReco(parts.reco, info, ctx);
	} else {
		parseBase(info.parts.base, info, ctx);
		parseReco(parts.reco, parts.base, info, [], ctx);
		parseAct(parts.act, info, ctx);
		parseUnknown(parts.unknown, info, ctx);
	}
	return info;
}

//#endregion
//#region src/bundle/manager.ts
var BundleManager = class {
	loader;
	watcher;
	root;
	delegate;
	changed;
	removed;
	watcherCtrl;
	duringFlush;
	flushResolve;
	needFlush;
	constructor(loader, watcher, root, delegate) {
		this.loader = loader;
		this.watcher = watcher;
		this.root = root;
		this.delegate = delegate;
		this.changed = /* @__PURE__ */ new Set();
		this.removed = /* @__PURE__ */ new Set();
		this.duringFlush = false;
		this.flushResolve = [];
		this.needFlush = false;
	}
	async load() {
		this.watcherCtrl?.stop();
		await this.delegate.reset();
		this.changed.clear();
		this.removed.clear();
		this.watcherCtrl = await this.watcher.watch(this.root, false, {
			filter: (file, isdir) => {
				return this.delegate.filterFile(file, isdir);
			},
			fileAdded: (file) => {
				this.changed.add(file);
				this.removed.delete(file);
				this.dispatchFlush();
			},
			fileChanged: (file) => {
				this.changed.add(file);
				this.removed.delete(file);
				this.dispatchFlush();
			},
			fileDeleted: (file) => {
				this.removed.add(file);
				this.changed.delete(file);
				this.dispatchFlush();
			}
		});
		await this.flush();
	}
	stop() {
		this.watcherCtrl?.stop();
	}
	async flush() {
		if (this.duringFlush) return new Promise((resolve) => {
			this.flushResolve.push(resolve);
		});
		this.duringFlush = true;
		this.needFlush = false;
		const changed = this.changed;
		const removed = this.removed;
		this.changed = /* @__PURE__ */ new Set();
		this.removed = /* @__PURE__ */ new Set();
		for (const file of removed) await this.delegate.deleteFile(relativePath(this.root, file), file);
		for (const file of changed) if (this.delegate.needContent(file)) {
			const content = await this.loader.get(file);
			if (typeof content === "string") await this.delegate.loadFile(relativePath(this.root, file), file, content);
			else await this.delegate.deleteFile(relativePath(this.root, file), file);
		} else await this.delegate.loadFile(relativePath(this.root, file), file);
		this.duringFlush = false;
		if (this.needFlush) setTimeout(() => {
			this.flush();
		}, 100);
		else {
			const resolves = this.flushResolve;
			this.flushResolve = [];
			process.nextTick(() => {
				for (const func of resolves) func();
			});
		}
	}
	dispatchFlush(timeout = 100) {
		if (this.needFlush) return;
		this.needFlush = true;
		setTimeout(() => {
			this.flush();
		}, timeout);
	}
};

//#endregion
//#region src/bundle/bundle.ts
var Bundle = class extends EventEmitter {
	maa;
	root;
	pipelineRoot;
	imageRoot;
	files;
	layer;
	manager;
	imageChangedTimer;
	get defaultPipelineRel() {
		return "default_pipeline.json";
	}
	get defaultPipelinePath() {
		return joinPath(this.root, this.defaultPipelineRel);
	}
	constructor(loader, watcher, maa, root) {
		super();
		this.maa = maa;
		this.root = root;
		this.pipelineRoot = joinPath(this.root, this.maa ? "tasks" : "pipeline");
		this.imageRoot = joinPath(this.root, this.maa ? "template" : "image");
		this.files = {};
		this.layer = new LayerInfo(loader, this.maa, this.root, "resource");
		this.manager = new BundleManager(loader, watcher, this.root, this);
	}
	async load() {
		await this.manager.load();
	}
	stop() {
		this.manager.stop();
	}
	async flush() {
		await this.manager.flush();
	}
	filterFile(file, isdir) {
		if (path.basename(file).startsWith(".")) return false;
		if (isdir) return file.startsWith(this.pipelineRoot) || file.startsWith(this.imageRoot) || file === this.root;
		else if (file.startsWith(this.pipelineRoot)) return file.endsWith(".json") || file.endsWith(".jsonc");
		else if (file.startsWith(this.imageRoot)) return file.endsWith(".png");
		else if (file === this.defaultPipelinePath) return true;
		return false;
	}
	needContent(file) {
		return file.endsWith(".json") || file.endsWith(".jsonc");
	}
	async reset() {
		this.files = {};
		this.layer.reset();
		this.emit("reset");
	}
	async loadFile(file, full, content) {
		if (!this.filterFile(full, false)) return;
		if (file.endsWith(".json") || file.endsWith(".jsonc")) {
			const changed = this.loadFileImpl(file, content);
			if (changed.length > 0) this.emit("taskChanged", [...new Set(changed)]);
		} else if (file.endsWith(".png")) {
			const imageFile = file.replaceAll(path.sep, "/").replace(this.maa ? "template/" : "image/", "");
			if (!this.layer.images.has(imageFile)) {
				this.layer.images.add(imageFile);
				this.dispatchImageChanged();
			}
		}
	}
	async deleteFile(file, full) {
		if (!this.filterFile(full, false)) return;
		if (file.endsWith(".json") || file.endsWith(".jsonc")) {
			const changed = this.deleteFileImpl(file);
			if (changed.length > 0) this.emit("taskChanged", [...new Set(changed)]);
		} else if (file.endsWith(".png")) {
			const imageFile = file.replaceAll(path.sep, "/").replace("image/", "");
			if (this.layer.images.delete(imageFile)) this.dispatchImageChanged();
		}
	}
	loadFileImpl(file, content) {
		const isDefault = file === this.defaultPipelineRel;
		const changed = [];
		changed.push(...this.deleteFileImpl(file));
		if (!content) return changed;
		this.files[file] = content;
		const full = joinPath(this.root, file);
		const tree = parseTreeWithoutParent(content);
		if (tree && tree.type === "object") for (const [key, obj, prop] of parseObject(tree)) {
			if (key.startsWith("$")) {
				if (key.startsWith("$__mpe")) this.layer.extraDecls.push({
					file: full,
					location: prop,
					type: "task.mpe_config"
				});
				continue;
			}
			let taskName = key;
			if (isDefault) taskName = "$" + taskName;
			this.layer.mutableTaskInfo(taskName).push({
				file: full,
				prop,
				data: obj,
				info: parseTask(obj, {
					maa: this.maa,
					file: full,
					task: prop,
					taskName
				}),
				obj: buildTree(obj)
			});
			this.layer.markDirty();
			changed.push(taskName);
		}
		return changed;
	}
	deleteFileImpl(file) {
		delete this.files[file];
		return this.layer.removeFile(joinPath(this.root, file));
	}
	dispatchImageChanged() {
		if (this.imageChangedTimer) clearTimeout(this.imageChangedTimer);
		this.imageChangedTimer = setTimeout(() => {
			this.emit("imageChanged");
		}, 100);
	}
};

//#endregion
//#region src/content/loader.ts
var FsContentLoader = class {
	async get(file) {
		try {
			return await fs.readFile(file, "utf8");
		} catch {
			return null;
		}
	}
};

//#endregion
//#region src/content/watch.ts
var FsContentWatcher = class {
	async watch(root, _isFile, delegate) {
		const watcher = chokidar.watch(root, { ignored: (file, stats) => {
			if (!stats) return false;
			return !delegate.filter(path.normalize(file), stats.isDirectory());
		} });
		watcher.on("add", (file) => {
			delegate.fileAdded(file);
		});
		watcher.on("change", (file) => {
			delegate.fileChanged(file);
		});
		watcher.on("unlink", (file) => {
			delegate.fileDeleted(file);
		});
		await new Promise((resolve) => {
			watcher.on("ready", () => {
				resolve();
			});
		});
		return { stop() {
			watcher.close();
		} };
	}
};

//#endregion
//#region src/diagnostic/interface.ts
function checkInterface(bundle) {
	const result = [];
	const layer = bundle.topLayer;
	if (layer.type === "interface") {
		const realTasks = new Set(layer.parent?.getTaskListNotUnique() ?? []);
		for (const ref of layer.mergedRefs) if (ref.type === "task.entry") {
			if (!realTasks.has(ref.target)) result.push({
				level: "error",
				file: ref.file,
				offset: ref.location.offset,
				length: ref.location.length,
				type: "int-unknown-entry-task",
				task: ref.target
			});
		}
		for (const decl of layer.mergedDecls) if (decl.type === "task.decl") {
			if (!realTasks.has(decl.task)) result.push({
				level: "error",
				file: decl.file,
				offset: decl.location.offset,
				length: decl.location.length,
				type: "int-override-unknown-task",
				task: decl.task
			});
		}
	}
	const ctrlDecls = bundle.info.decls.filter((decl) => decl.type === "interface.controller");
	const ctrls = /* @__PURE__ */ new Map();
	for (const decl of ctrlDecls) if (ctrls.has(decl.name)) {
		const prev = ctrls.get(decl.name);
		result.push({
			level: "error",
			file: decl.file,
			offset: decl.location.offset,
			length: decl.location.length,
			type: "int-conflict-controller",
			ctrl: decl.name,
			previous: {
				file: prev.file,
				offset: prev.location.offset,
				length: prev.location.length
			}
		});
	} else ctrls.set(decl.name, decl);
	const resDecls = bundle.info.decls.filter((decl) => decl.type === "interface.resource");
	const ress = /* @__PURE__ */ new Map();
	for (const decl of resDecls) if (ress.has(decl.name)) {
		const prev = ress.get(decl.name);
		result.push({
			level: "error",
			file: decl.file,
			offset: decl.location.offset,
			length: decl.location.length,
			type: "int-conflict-resource",
			res: decl.name,
			previous: {
				file: prev.file,
				offset: prev.location.offset,
				length: prev.location.length
			}
		});
	} else ress.set(decl.name, decl);
	const optDecls = bundle.info.decls.filter((decl) => decl.type === "interface.option");
	const options = /* @__PURE__ */ new Map();
	for (const decl of optDecls) if (options.has(decl.name)) {
		const prev = options.get(decl.name);
		result.push({
			level: "error",
			file: decl.file,
			offset: decl.location.offset,
			length: decl.location.length,
			type: "int-conflict-option",
			option: decl.name,
			previous: {
				file: prev.file,
				offset: prev.location.offset,
				length: prev.location.length
			}
		});
	} else {
		options.set(decl.name, decl);
		if (!decl.optionType || decl.optionType === "select" || decl.optionType === "switch") {
			const caseDecls = bundle.info.decls.filter((decl2) => decl2.type === "interface.case" && decl2.option === decl.name);
			const cases = /* @__PURE__ */ new Map();
			for (const decl2 of caseDecls) if (cases.has(decl2.name)) {
				const prev = cases.get(decl2.name);
				result.push({
					level: "error",
					file: decl2.file,
					offset: decl2.location.offset,
					length: decl2.location.length,
					type: "int-conflict-case",
					option: decl.name,
					case: decl2.name,
					previous: {
						file: prev.file,
						offset: prev.location.offset,
						length: prev.location.length
					}
				});
			} else cases.set(decl2.name, decl2);
			const caseRefs = bundle.info.refs.filter((ref) => ref.type === "interface.case").filter((ref) => ref.option === decl.name);
			for (const ref of caseRefs) if (!cases.has(ref.target)) result.push({
				level: "error",
				file: ref.file,
				offset: ref.location.offset,
				length: ref.location.length,
				type: "int-unknown-case",
				option: decl.name,
				case: ref.target
			});
			if (decl.optionType === "switch") {
				let missingYes = true;
				let missingNo = true;
				for (const [name, decl2] of cases) if (name === "Yes") missingYes = false;
				else if (name === "No") missingNo = false;
				else if (name.toLowerCase() === "yes") {
					missingYes = false;
					result.push({
						level: "warning",
						file: decl2.file,
						offset: decl2.location.offset,
						length: decl2.location.length,
						type: "int-switch-should-fixed"
					});
				} else if (name.toLowerCase() === "no") {
					missingNo = false;
					result.push({
						level: "warning",
						file: decl2.file,
						offset: decl2.location.offset,
						length: decl2.location.length,
						type: "int-switch-should-fixed"
					});
				} else result.push({
					level: "error",
					file: decl2.file,
					offset: decl2.location.offset,
					length: decl2.location.length,
					type: "int-switch-name-invalid"
				});
				if (missingYes || missingNo) result.push({
					level: "error",
					file: decl.file,
					offset: decl.location.offset,
					length: decl.location.length,
					type: "int-switch-missing",
					option: decl.name,
					missingYes,
					missingNo
				});
			}
		}
	}
	for (const ref of bundle.info.refs) if (ref.type === "interface.controller") {
		if (!ctrls.has(ref.target)) result.push({
			level: "error",
			file: ref.file,
			offset: ref.location.offset,
			length: ref.location.length,
			type: "int-unknown-controller",
			ctrl: ref.target
		});
	} else if (ref.type === "interface.resource") {
		if (!ress.has(ref.target)) result.push({
			level: "error",
			file: ref.file,
			offset: ref.location.offset,
			length: ref.location.length,
			type: "int-unknown-resource",
			res: ref.target
		});
	} else if (ref.type === "interface.option") {
		if (!options.has(ref.target)) result.push({
			level: "error",
			file: ref.file,
			offset: ref.location.offset,
			length: ref.location.length,
			type: "int-unknown-option",
			option: ref.target
		});
		if (ref.preset) {
			const optDecl = optDecls.find((decl) => decl.name === ref.target);
			if (optDecl) switch (optDecl.optionType ?? "select") {
				case "select":
				case "switch":
					if (!isString(ref.preset)) result.push({
						level: "error",
						file: ref.file,
						offset: ref.preset.offset,
						length: ref.preset.length,
						type: "int-preset-type-error",
						option: ref.target,
						expected: "string"
					});
					break;
				case "checkbox":
					if (ref.preset.type !== "array") result.push({
						level: "error",
						file: ref.file,
						offset: ref.preset.offset,
						length: ref.preset.length,
						type: "int-preset-type-error",
						option: ref.target,
						expected: "array"
					});
					break;
				case "input":
					if (ref.preset.type !== "object") result.push({
						level: "error",
						file: ref.file,
						offset: ref.preset.offset,
						length: ref.preset.length,
						type: "int-preset-type-error",
						option: ref.target,
						expected: "object"
					});
					break;
			}
		}
	}
	return result;
}

//#endregion
//#region src/utils/helper.ts
function filterDeclRef(infos, offset) {
	return infos.filter((info) => info.location.offset <= offset && info.location.offset + info.location.length >= offset);
}
function findDeclRef(infos, offset) {
	let result = null;
	for (const info of filterDeclRef(infos, offset)) if (result === null) result = info;
	else if (info.location.offset > result.location.offset) result = info;
	else if (info.location.offset < result.location.offset) continue;
	else if (info.location.length < result.location.length) result = info;
	return result;
}
function extractTaskRef(r) {
	if (r.type === "task.anchor" || r.type === "task.reco" || r.type === "task.color_filter" || r.type === "task.entry") return r.target;
	else if (r.type === "task.next" || r.type === "task.roi" || r.type === "task.target") {
		if (r.attrs.attrs.Anchor) return null;
		if (r.type === "task.roi" && r.prevRef) return null;
		return r.target;
	} else return null;
}
function isAnchorRef(r) {
	return (r.type === "task.next" || r.type === "task.roi" || r.type === "task.target") && !!r.attrs.attrs.Anchor;
}
function findMaaDeclRef(infos, offset) {
	for (const info of infos) if (offset >= info.offset && offset <= info.offset + info.length) return info;
	return null;
}

//#endregion
//#region src/diagnostic/task.ts
function checkTask(bundle) {
	const result = [];
	for (const layer of bundle.allLayers) {
		for (const [name, taskInfos] of Object.entries(layer.tasks)) {
			if (taskInfos.length > 0 && layer.type !== "interface") for (const taskInfo of taskInfos.slice(1)) result.push({
				level: "error",
				file: taskInfo.file,
				offset: taskInfo.prop.offset,
				length: taskInfo.prop.length,
				type: "conflict-task",
				task: name,
				previous: {
					file: taskInfos[0].file,
					offset: taskInfos[0].prop.offset,
					length: taskInfos[0].prop.length
				}
			});
			if (!bundle.maa) for (const taskInfo of taskInfos) {
				const existsNext = /* @__PURE__ */ new Set();
				const refs = taskInfo.info.refs.filter((ref) => ref.type === "task.next" && !ref.attrs.attrs.Anchor);
				refs.sort((a, b) => a.location.offset - b.location.offset);
				for (const ref of refs) if (existsNext.has(ref.target)) result.push({
					level: "error",
					file: ref.file,
					offset: ref.location.offset,
					length: ref.location.length,
					type: "duplicate-next",
					task: ref.target
				});
				else existsNext.add(ref.target);
			}
		}
		const decls = layer.mergedDecls;
		for (const decl of decls) if (decl.type === "task.mpe_config") result.push({
			level: "warning",
			file: decl.file,
			offset: decl.location.offset,
			length: decl.location.length,
			type: "mpe-config"
		});
		const refs = layer.mergedRefs;
		const tasks = new Set(layer.getTaskListNotUnique());
		const anchors = new Set(layer.getAnchorList().map(([anchor]) => anchor));
		const images = new Set(layer.getImageListNotUnique());
		const imageFolders = layer.getImageFolders();
		for (const ref of refs) {
			const task = extractTaskRef(ref);
			if (task !== null) {
				if (!tasks.has(task) && !(task === "" && ref.type === "task.anchor")) {
					let offset = ref.location.offset;
					let length = ref.location.length;
					if (ref.type === "task.next" && ref.attrs.offset > 0) {
						offset = ref.location.offset + ref.attrs.offset + 1;
						length = ref.location.length - ref.attrs.offset - 2;
					}
					result.push({
						level: "error",
						file: ref.file,
						offset,
						length,
						type: "unknown-task",
						task
					});
				}
				if (ref.type === "task.color_filter") {
					const { reco } = layer.getTaskBriefInfo(ref.target);
					if (reco !== "ColorMatch") result.push({
						level: "error",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "color-filter-invalid",
						task,
						reco: reco ?? "DirectHit"
					});
				}
			} else if (ref.type === "task.template") {
				let imagePath = ref.target;
				let isFolder = false;
				if (!bundle.maa && !imagePath.endsWith(".png")) {
					const norm = normalizeImageFolder(imagePath);
					if (imageFolders.has(norm)) isFolder = true;
					else {
						result.push({
							level: "warning",
							file: ref.file,
							offset: ref.location.offset,
							length: ref.location.length,
							type: "dynamic-image"
						});
						continue;
					}
				}
				if (imagePath.includes("\\")) {
					result.push({
						level: "warning",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "image-path-back-slash"
					});
					imagePath = imagePath.replaceAll("\\", "/");
				}
				if (imagePath.startsWith("./")) {
					result.push({
						level: "warning",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "image-path-dot-slash"
					});
					imagePath = imagePath.replace("./", "");
				}
				if (bundle.maa && !imagePath.endsWith(".png")) {
					result.push({
						level: "warning",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "image-path-missing-png"
					});
					imagePath = imagePath + ".png";
				}
				if (isFolder) continue;
				if (!images.has(imagePath)) {
					let found = false;
					if (bundle.maa) {
						const suffix = "/" + imagePath;
						for (const image of images) if (image.endsWith(suffix)) {
							found = true;
							break;
						}
					}
					if (!found) result.push({
						level: "error",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "unknown-image",
						image: ref.target
					});
				}
			} else if (isAnchorRef(ref)) {
				if (!anchors.has(ref.target)) result.push({
					level: "error",
					file: ref.file,
					offset: ref.location.offset + ref.attrs.offset + 1,
					length: ref.location.length - ref.attrs.offset - 2,
					type: "unknown-anchor",
					anchor: ref.target
				});
			} else if (ref.type === "task.locale") {
				const infos = bundle.langBundle.queryKey(ref.target);
				if (!infos.find((info) => !!info)) result.push({
					level: "error",
					file: ref.file,
					offset: ref.location.offset,
					length: ref.location.length,
					type: "unknown-locale",
					locale: ref.target
				});
				else {
					const missingLangs = [];
					for (const [idx, info] of infos.entries()) if (!info) missingLangs.push(bundle.langBundle.langs[idx].name);
					if (missingLangs.length > 0) result.push({
						level: "error",
						file: ref.file,
						offset: ref.location.offset,
						length: ref.location.length,
						type: "missing-locale",
						locale: ref.target,
						langs: missingLangs
					});
				}
			}
			if ((ref.type === "task.next" || ref.type === "task.roi" || ref.type === "task.target") && ref.attrs.unknown.length > 0) for (const [attr, offset, length] of ref.attrs.unknown) result.push({
				level: "error",
				file: ref.file,
				offset: ref.location.offset + 2 + offset,
				length: length - 2,
				type: "unknown-attr",
				attr
			});
		}
	}
	return result;
}

//#endregion
//#region src/diagnostic/message.ts
async function buildDiagnosticMessage(root, diag, evalPos, _option) {
	const buildPos = async (loc) => {
		const [line, column] = await evalPos(loc.file, loc.offset);
		return `${relativePath(root, loc.file)}:${line}:${column}`;
	};
	const start = await evalPos(diag.file, diag.offset);
	const end = await evalPos(diag.file, diag.offset + diag.length);
	const buildBrief = async () => {
		switch (diag.type) {
			case "conflict-task": return t("maa.pipeline.error.conflict-task", diag.task, await buildPos(diag.previous));
			case "duplicate-next": return t("maa.pipeline.error.duplicate-next", diag.task);
			case "unknown-task": return t("maa.pipeline.error.unknown-task", diag.task);
			case "color-filter-invalid": return t("maa.pipeline.error.color-filter-invalid", diag.task, diag.reco);
			case "dynamic-image": return t("maa.pipeline.warning.image-path-dynamic");
			case "image-path-back-slash": return t("maa.pipeline.warning.image-path-backslash");
			case "image-path-dot-slash": return t("maa.pipeline.warning.image-path-dot-slash");
			case "image-path-missing-png": return t("maa.pipeline.warning.image-path-missing-png");
			case "unknown-image": return t("maa.pipeline.error.unknown-image", diag.image);
			case "unknown-anchor": return t("maa.pipeline.error.unknown-anchor", diag.anchor);
			case "unknown-attr": return t("maa.pipeline.error.unknown-attr", diag.attr);
			case "unknown-locale": return t("maa.pipeline.error.unknown-locale", diag.locale);
			case "missing-locale": return t("maa.pipeline.error.missing-locale", diag.locale, diag.langs.join(", "));
			case "mpe-config": return t("maa.pipeline.warning.mpe-config");
			case "int-conflict-controller": return t("maa.pipeline.error.conflict-controller", diag.ctrl, await buildPos(diag.previous));
			case "int-unknown-controller": return t("maa.pipeline.error.unknown-controller", diag.ctrl);
			case "int-conflict-resource": return t("maa.pipeline.error.conflict-resource", diag.res, await buildPos(diag.previous));
			case "int-unknown-resource": return t("maa.pipeline.error.unknown-resource", diag.res);
			case "int-conflict-option": return t("maa.pipeline.error.conflict-option", diag.option, await buildPos(diag.previous));
			case "int-unknown-option": return t("maa.pipeline.error.unknown-option", diag.option);
			case "int-conflict-case": return t("maa.pipeline.error.conflict-case", diag.case, diag.option, await buildPos(diag.previous));
			case "int-unknown-case": return t("maa.pipeline.error.unknown-case", diag.case, diag.option);
			case "int-switch-name-invalid": return t("maa.pipeline.error.switch-name-invalid");
			case "int-switch-missing": if (diag.missingYes && diag.missingNo) return t("maa.pipeline.error.switch-missing-all");
			else if (diag.missingYes) return t("maa.pipeline.error.switch-missing-yes");
			else return t("maa.pipeline.error.switch-missing-no");
			case "int-switch-should-fixed": return t("maa.pipeline.warning.switch-name-should-fixed");
			case "int-preset-type-error": return t("maa.pipeline.error.preset-type-error", diag.option, diag.expected);
			case "int-unknown-entry-task": return t("maa.pipeline.error.unknown-entry-task", diag.task);
			case "int-override-unknown-task": return t("maa.pipeline.error.override-unknown-task", diag.task);
		}
		return `unknown diagnostic: ${JSON.stringify(diag)}`;
	};
	return [
		start,
		end,
		await buildBrief()
	];
}

//#endregion
//#region src/diagnostic/diagnostic.ts
function performDiagnostic(bundle, option) {
	const result = [];
	result.push(...checkTask(bundle));
	result.push(...checkInterface(bundle));
	return result.filter((diag) => !option.ignoreTypes?.includes(diag.type)).map((diag) => {
		if (option.errorTypes?.includes(diag.type)) {
			const newDiag = { ...diag };
			newDiag.level = "error";
			return newDiag;
		} else return diag;
	});
}

//#endregion
//#region src/content/json.ts
var ContentJson = class {
	loader;
	watcher;
	file;
	changed;
	node;
	object;
	dirty;
	watcherCtrl;
	duringFlush;
	flushResolve;
	needFlush;
	constructor(loader, watcher, file, changed) {
		this.loader = loader;
		this.watcher = watcher;
		this.file = file;
		this.changed = changed;
		this.dirty = true;
		this.duringFlush = false;
		this.flushResolve = [];
		this.needFlush = false;
		this.load();
	}
	async load() {
		this.watcherCtrl?.stop();
		this.dirty = true;
		this.watcherCtrl = await this.watcher.watch(this.file, true, {
			filter: (_file, _isdir) => {
				return true;
			},
			fileAdded: (_file) => {
				this.dirty = true;
				this.dispatchFlush();
			},
			fileChanged: (_file) => {
				this.dirty = true;
				this.dispatchFlush();
			},
			fileDeleted: (_file) => {
				this.dirty = true;
				this.dispatchFlush();
			}
		});
		await this.flush();
	}
	stop() {
		this.watcherCtrl?.stop();
	}
	async flush() {
		if (this.duringFlush) return new Promise((resolve) => {
			this.flushResolve.push(resolve);
		});
		this.duringFlush = true;
		this.needFlush = false;
		if (this.dirty) {
			const content = await this.loader.get(this.file);
			if (typeof content === "string") this.node = parseTreeWithoutParent(content);
			else this.node = void 0;
			if (this.node) this.object = buildTree(this.node);
			else this.object = void 0;
			await this.changed(this.node, this.object);
			this.dirty = false;
		}
		const resolves = this.flushResolve;
		this.flushResolve = [];
		this.duringFlush = false;
		process.nextTick(() => {
			for (const func of resolves) func();
		});
		if (this.needFlush) setTimeout(() => {
			this.flush();
		}, 100);
	}
	dispatchFlush(timeout = 100) {
		if (this.needFlush) return;
		this.needFlush = true;
		setTimeout(() => {
			this.flush();
		}, timeout);
	}
};

//#endregion
//#region src/parser/interface/ctrlRef.ts
function parseCtrlRef(node, info, ctx) {
	const refs = [];
	for (const obj of parseArray(node)) if (isString(obj)) {
		info.refs.push({
			file: ctx.file,
			location: obj,
			type: "interface.controller",
			target: obj.value
		});
		refs.push(obj.value);
	}
	return refs;
}

//#endregion
//#region src/parser/interface/import.ts
function parseImport(node, info, ctx) {
	if (isString(node)) info.refs.push({
		file: ctx.file,
		location: node,
		type: "interface.import_path",
		target: node.value
	});
}

//#endregion
//#region src/parser/interface/keys.ts
const locKeys = [
	"label",
	"icon",
	"description",
	"title",
	"contact",
	"license",
	"welcome"
];

//#endregion
//#region src/parser/interface/language.ts
function parseLanguage(node, info, ctx) {
	for (const [key, obj] of parseObject(node)) if (isString(obj)) {
		info.decls.push({
			file: ctx.file,
			location: obj,
			type: "interface.language",
			name: key,
			path: obj.value
		});
		info.refs.push({
			file: ctx.file,
			location: obj,
			type: "interface.language_path",
			target: obj.value
		});
	}
}

//#endregion
//#region src/parser/interface/optionRef.ts
function parseOptionRef(node, info, ctx, trace) {
	for (const obj of parseArray(node)) if (isString(obj)) info.refs.push({
		file: ctx.file,
		location: obj,
		type: "interface.option",
		target: obj.value,
		trace: {
			name: obj.value,
			...trace
		}
	});
}

//#endregion
//#region src/parser/interface/override.ts
function parseOverride(node, info, ctx) {
	for (const [key, obj, prop] of parseObject(node)) {
		if (key.startsWith("$")) {
			if (key.startsWith("$__mpe")) info.layer.extraDecls.push({
				file: ctx.file,
				location: prop,
				type: "task.mpe_config"
			});
			continue;
		}
		info.layer.mutableTaskInfo(key).push({
			file: ctx.file,
			prop,
			data: obj,
			info: parseTask(obj, {
				maa: ctx.maa,
				file: ctx.file,
				task: prop,
				taskName: key
			}),
			obj: buildTree(obj)
		});
		info.layer.markDirty();
	}
}

//#endregion
//#region src/parser/interface/case.ts
function parseCase(node, info, option, ctx) {
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) info.decls.push({
				file: ctx.file,
				location: obj,
				type: "interface.case",
				name: obj.value,
				option
			});
			break;
		case "option":
			parseOptionRef(obj, info, ctx, {
				from: "option",
				origin: option
			});
			break;
		case "pipeline_override":
			parseOverride(obj, info, ctx);
			break;
	}
}
function parseCases(node, info, option, ctx) {
	for (const obj of parseArray(node)) parseCase(obj, info, option, ctx);
}

//#endregion
//#region src/parser/interface/input.ts
function isPipelineType(type) {
	return [
		"string",
		"int",
		"bool"
	].includes(type);
}
function parseInput(node, info, option, ctx) {
	let loc = null;
	const decl = {
		type: "interface.input",
		name: "",
		option
	};
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) {
				loc = obj;
				decl.name = obj.value;
			}
			break;
		case "pipeline_type":
			if (isString(obj) && isPipelineType(obj.value)) decl.cast = obj.value;
			break;
	}
	if (loc) {
		info.decls.push({
			file: ctx.file,
			location: loc,
			...decl
		});
		return decl.name;
	} else return null;
}
function parseInputs(node, info, option, ctx) {
	const names = [];
	for (const obj of parseArray(node)) {
		const name = parseInput(obj, info, option, ctx);
		if (name) names.push(name);
	}
	return names;
}

//#endregion
//#region src/parser/interface/resRef.ts
function parseResRef(node, info, ctx) {
	for (const obj of parseArray(node)) if (isString(obj)) info.refs.push({
		file: ctx.file,
		location: obj,
		type: "interface.resource",
		target: obj.value
	});
}

//#endregion
//#region src/parser/interface/option.ts
function parseInputRef(node, info, option, names, ctx) {
	if (isString(node)) for (const [name, re] of names) for (const occur of node.value.matchAll(re)) info.refs.push({
		file: ctx.file,
		location: node,
		type: "interface.input",
		target: name,
		option,
		offset: occur.index
	});
	else if (node.type === "array") for (const obj of parseArray(node)) parseInputRef(obj, info, option, names, ctx);
	else if (node.type === "object") for (const [, obj] of parseObject(node)) parseInputRef(obj, info, option, names, ctx);
}
function parseOptionSec(node, info, option, ctx) {
	let type = void 0;
	let inputNames = [];
	let overrideNode = null;
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "controller":
			parseCtrlRef(obj, info, ctx);
			break;
		case "resource":
			parseResRef(obj, info, ctx);
			break;
		case "type":
			if (isString(obj)) {
				if ([
					"select",
					"checkbox",
					"switch",
					"input"
				].includes(obj.value)) type = obj.value;
			}
			break;
		case "cases":
			parseCases(obj, info, option, ctx);
			break;
		case "inputs":
			inputNames = parseInputs(obj, info, option, ctx);
			break;
		case "pipeline_override":
			overrideNode = obj;
			break;
		case "default_case":
			if (isString(obj)) info.refs.push({
				file: ctx.file,
				location: obj,
				type: "interface.case",
				target: obj.value,
				option
			});
			else for (const sub of parseArray(obj)) if (isString(sub)) info.refs.push({
				file: ctx.file,
				location: sub,
				type: "interface.case",
				target: sub.value,
				option
			});
			break;
	}
	if (overrideNode) {
		const names = [];
		for (const name of inputNames) names.push([name, new RegExp("\\{" + name + "\\}", "g")]);
		parseInputRef(overrideNode, info, option, names, ctx);
		parseOverride(overrideNode, info, ctx);
	}
	return type;
}
function parseOption(node, info, ctx) {
	for (const [key, obj, prop] of parseObject(node)) {
		const type = parseOptionSec(obj, info, key, ctx);
		info.decls.push({
			file: ctx.file,
			location: prop,
			type: "interface.option",
			name: key,
			optionType: type
		});
	}
}

//#endregion
//#region src/parser/interface/path.ts
function parseSingle(node, info, ctx) {
	if (isString(node)) {
		let target = node.value;
		if (target.startsWith("{PROJECT_DIR}")) target = target.substring(14);
		info.refs.push({
			file: ctx.file,
			location: node,
			type: "interface.resource_path",
			target
		});
		return target;
	} else return "";
}
function parsePath(node, info, ctx) {
	const result = [];
	if (node.type !== "array") result.push(parseSingle(node, info, ctx));
	else for (const obj of parseArray(node)) result.push(parseSingle(obj, info, ctx));
	return result.filter((res) => res !== "");
}

//#endregion
//#region src/parser/interface/preset.ts
function parsePresetOption(node, info, ctx, name) {
	for (const [key, obj, prop] of parseObjectFlex(node)) {
		info.refs.push({
			file: ctx.file,
			location: prop,
			type: "interface.option",
			target: key,
			trace: {
				name: key,
				from: "preset",
				origin: name
			},
			preset: obj ?? void 0
		});
		if (!obj) continue;
		if (isString(obj)) info.refs.push({
			file: ctx.file,
			location: obj,
			type: "interface.case",
			option: key,
			target: obj.value
		});
		else if (obj.type === "array") {
			for (const val of parseArray(obj)) if (isString(val)) info.refs.push({
				file: ctx.file,
				location: val,
				type: "interface.case",
				target: val.value,
				option: key
			});
		} else for (const [inputKey, _, inputProp] of parseObjectFlex(obj)) info.refs.push({
			file: ctx.file,
			location: inputProp,
			type: "interface.input",
			target: inputKey,
			option: key
		});
	}
}
function parsePresetTask(node, info, ctx) {
	let name = "";
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) {
				name = obj.value;
				info.refs.push({
					file: ctx.file,
					location: obj,
					type: "interface.task",
					target: obj.value
				});
			}
			break;
	}
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "option":
			parsePresetOption(obj, info, ctx, name);
			break;
	}
}
function parsePresetSingle(node, info, ctx) {
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) info.decls.push({
				file: ctx.file,
				location: obj,
				type: "interface.preset",
				name: obj.value
			});
			break;
		case "task": for (const sub of parseArray(obj)) parsePresetTask(sub, info, ctx);
	}
}
function parsePreset(node, info, ctx) {
	for (const obj of parseArray(node)) parsePresetSingle(obj, info, ctx);
}

//#endregion
//#region src/parser/interface/interface.ts
function parseController(node, info, ctx) {
	let loc = null;
	const decl = {
		type: "interface.controller",
		name: "",
		attachs: []
	};
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) {
				loc = obj;
				decl.name = obj.value;
			}
			break;
		case "attach_resource_path":
			decl.attachs = parsePath(obj, info, ctx);
			break;
	}
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "option":
			parseOptionRef(obj, info, ctx, {
				from: "controller",
				origin: decl.name
			});
			break;
	}
	if (loc) info.decls.push({
		file: ctx.file,
		location: loc,
		...decl
	});
}
function parseResource(node, info, ctx) {
	let loc = null;
	const decl = {
		type: "interface.resource",
		name: "",
		paths: []
	};
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) {
				loc = obj;
				decl.name = obj.value;
			}
			break;
		case "path":
			decl.paths = parsePath(obj, info, ctx);
			break;
		case "controller":
			decl.controller = parseCtrlRef(obj, info, ctx);
			break;
	}
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "option":
			parseOptionRef(obj, info, ctx, {
				from: "resource",
				origin: decl.name
			});
			break;
	}
	if (loc) info.decls.push({
		file: ctx.file,
		location: loc,
		...decl
	});
}
function parseTaskSec(node, info, ctx) {
	let name = "";
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "name":
			if (isString(obj)) {
				name = obj.value;
				info.decls.push({
					file: ctx.file,
					location: obj,
					type: "interface.task",
					name: obj.value
				});
			}
			break;
		case "resource":
			parseResRef(obj, info, ctx);
			break;
		case "controller":
			parseCtrlRef(obj, info, ctx);
			break;
		case "pipeline_override":
			parseOverride(obj, info, ctx);
			break;
	}
	for (const [key, obj] of parseObject(node)) switch (key) {
		case "entry":
			if (isString(obj)) {
				info.refs.push({
					file: ctx.file,
					location: obj,
					type: "interface.task_entry",
					target: obj.value,
					task: name
				});
				info.layer.extraRefs.push({
					file: ctx.file,
					location: obj,
					type: "task.entry",
					target: obj.value
				});
			}
			break;
		case "option":
			parseOptionRef(obj, info, ctx, {
				from: "task",
				origin: name
			});
			break;
	}
}
function parseLocalization(node, info, ctx) {
	if (node.type === "object") for (const [key, obj] of parseObject(node)) if (locKeys.includes(key) && isString(obj)) {
		if (obj.value.startsWith("$")) info.layer.extraRefs.push({
			file: ctx.file,
			location: obj,
			type: "task.locale",
			target: obj.value.substring(1)
		});
		else if (obj.value.length > 0) info.layer.extraRefs.push({
			file: ctx.file,
			location: obj,
			type: "task.can_locale",
			target: obj.value
		});
	} else parseLocalization(obj, info, ctx);
	else if (node.type === "array") for (const obj of parseArray(node)) parseLocalization(obj, info, ctx);
}
function parseInterface(node, info, ctx) {
	for (const [key, obj] of parseObject(node)) {
		if (ctx.import && ![
			"option",
			"task",
			"preset"
		].includes(key)) continue;
		switch (key) {
			case "languages":
				parseLanguage(obj, info, ctx);
				break;
			case "controller":
				for (const sub of parseArray(obj)) parseController(sub, info, ctx);
				break;
			case "resource":
				for (const sub of parseArray(obj)) parseResource(sub, info, ctx);
				break;
			case "task":
				for (const sub of parseArray(obj)) parseTaskSec(sub, info, ctx);
				break;
			case "option":
				parseOption(obj, info, ctx);
				break;
			case "global_option":
				parseOptionRef(obj, info, ctx, {
					from: "global",
					origin: ""
				});
				break;
			case "import":
				for (const sub of parseArray(obj)) parseImport(sub, info, ctx);
				break;
			case "preset":
				parsePreset(obj, info, ctx);
				break;
		}
	}
	parseLocalization(node, info, ctx);
}

//#endregion
//#region src/interface/language.ts
var LanguageBundle = class extends EventEmitter {
	loader;
	watcher;
	root;
	langs;
	constructor(loader, watcher, root) {
		super();
		this.loader = loader;
		this.watcher = watcher;
		this.root = root;
		this.langs = [];
	}
	stop() {
		for (const lang of this.langs) lang.content.stop();
	}
	async flush() {
		await Promise.all(this.langs.map((lang) => lang.content.flush()));
	}
	async update(config) {
		if (JSON.stringify(this.langs.map((info) => [info.name, info.file])) === JSON.stringify(config)) return true;
		this.stop();
		this.langs = config.map(([name, file], idx) => ({
			name,
			file,
			content: new ContentJson(this.loader, this.watcher, joinPath(this.root, file), async () => {
				await this.rebuildIndex(idx);
			}),
			entries: [],
			decls: [],
			refs: []
		}));
		await Promise.all(this.langs.map((lang) => lang.content.load()));
	}
	async rebuildIndex(idx) {
		const lang = this.langs[idx];
		const full = joinPath(this.root, lang.file);
		lang.entries = [];
		lang.decls = [];
		lang.refs = [];
		for (const [key, obj, prop] of parseObject(lang.content.node)) if (isString(obj)) {
			lang.entries.push({
				key,
				keyNode: prop,
				value: obj.value,
				valueNode: obj
			});
			lang.decls.push({
				location: prop,
				file: full,
				type: "task.locale",
				key,
				value: obj.value,
				valueNode: obj
			});
			lang.refs.push({
				location: obj,
				file: full,
				type: "task.locale_text",
				target: obj.value
			});
		}
		this.emit("localeChanged");
	}
	allKeys() {
		const keys = this.langs.map((lang) => lang.entries.map((entry) => entry.key)).flat();
		return [...new Set(keys)];
	}
	queryName(name) {
		const idx = this.langs.findIndex((lang) => lang.name === name);
		return idx === -1 ? 0 : idx;
	}
	queryKey(key) {
		const result = [];
		for (const lang of this.langs) {
			const info = lang.entries.find((info) => info.key === key);
			result.push(info ?? null);
		}
		return result;
	}
	addPair(key, value, indent = "    ") {
		const keys = this.allKeys();
		if (keys.includes(key)) return [];
		const row = JSON.stringify(key) + ": " + JSON.stringify(value);
		const result = [];
		if (keys.length === 0) return this.langs.map((lang) => {
			return {
				type: "replace",
				file: joinPath(this.root, lang.file),
				content: `{
${indent}${row},
}
`
			};
		});
		else {
			let insertIndex = keys.findIndex((val) => val.localeCompare(key) > 0);
			if (insertIndex === -1) insertIndex = keys.length;
			const upper = keys.slice(0, insertIndex);
			const lower = keys.slice(insertIndex);
			for (const lang of this.langs) {
				let found = false;
				for (const upKey of upper.toReversed()) {
					const anchor = lang.entries.find((entry) => entry.key === upKey);
					if (!anchor) continue;
					result.push({
						type: "insert",
						file: joinPath(this.root, lang.file),
						content: `,\n${indent}${row}`,
						offset: anchor.valueNode.offset + anchor.valueNode.length
					});
					found = true;
					break;
				}
				if (found) continue;
				for (const loKey of lower) {
					const anchor = lang.entries.find((entry) => entry.key === loKey);
					if (!anchor) continue;
					result.push({
						type: "insert",
						file: joinPath(this.root, lang.file),
						content: `${row},\n${indent}`,
						offset: anchor.keyNode.offset
					});
					found = true;
					break;
				}
				if (found) continue;
				result.push({
					type: "replace",
					file: joinPath(this.root, lang.file),
					content: `{
${indent}${row},
}
`
				});
			}
		}
		return result;
	}
};

//#endregion
//#region src/interface/interface.ts
var MaaEvalDelegateImpl = class extends MaaEvalDelegate {
	intBundle;
	constructor(intBundle) {
		super(new MaaErrorDelegate());
		this.intBundle = intBundle;
	}
	query(task) {
		const topLayer = this.intBundle.topLayer;
		if (!topLayer) return [];
		const infos = topLayer.getTask(task, false);
		infos.reverse();
		return infos.map(({ layer, infos }) => {
			const info = infos[0];
			const match = /resource\/global\/(.+)\//.exec(layer.root);
			const anchor = match ? match[1] : "Official";
			return [info.obj, anchor];
		});
	}
};
var InterfaceBundle = class extends EventEmitter {
	maa;
	root;
	file;
	content;
	info;
	importFiles;
	imports;
	activeController;
	activeResource;
	paths;
	bundles;
	langBundle;
	eval;
	set evalErrorDelegate(delegate) {
		this.eval.error = delegate;
	}
	constructor(loader, watcher, maa, root, file = "interface.json") {
		super();
		this.maa = maa;
		this.root = root;
		this.file = joinPath(this.root, file);
		this.content = new ContentJson(loader, watcher, this.file, () => {
			this.removeFile(this.file);
			if (this.content.node) parseInterface(this.content.node, this.info, {
				maa: this.maa,
				file: this.file,
				import: false
			});
			this.emit("interfaceChanged");
		});
		this.info = {
			decls: [],
			refs: [],
			layer: new LayerInfo(loader, this.maa, this.root, "interface")
		};
		this.activeController = "";
		this.activeResource = "";
		this.paths = [];
		this.bundles = [];
		this.langBundle = new LanguageBundle(loader, watcher, this.root);
		this.importFiles = [];
		this.imports = [];
		this.eval = new MaaEvalDelegateImpl(this);
		this.on("interfaceChanged", () => {
			this.updatePaths();
			this.updateLangs();
			this.updateImports();
		});
		this.on("activeChanged", () => {
			this.updatePaths();
		});
		this.on("importChanged", async () => {
			await Promise.all(this.imports.map((content) => content.load()));
		});
		this.on("pathChanged", async () => {
			let prev = void 0;
			for (const bundle of this.bundles) {
				bundle.layer.parent = prev;
				prev = bundle.layer;
			}
			this.info.layer.parent = prev;
			await Promise.all(this.bundles.map((bundle) => bundle.load()));
			this.emit("bundleReloaded");
			this.emit("switchActiveFinished");
		});
		this.on("bundleReloaded", () => {
			for (const bundle of this.bundles) {
				bundle.on("reset", () => {
					this.emit("pipelineChanged");
				});
				bundle.on("taskChanged", () => {
					this.emit("pipelineChanged");
				});
				bundle.on("imageChanged", () => {
					this.emit("pipelineChanged");
				});
			}
		});
		this.langBundle.on("localeChanged", () => {
			const restDecls = this.info.layer.extraDecls.filter((decl) => decl.type !== "task.locale");
			this.info.layer.extraDecls = [...restDecls, ...this.langBundle.langs.map((lang) => lang.decls).flat()];
			const restRefs = this.info.layer.extraRefs.filter((ref) => ref.type !== "task.locale_text");
			this.info.layer.extraRefs = [...restRefs, ...this.langBundle.langs.map((lang) => lang.refs).flat()];
			this.emit("localeChanged");
		});
	}
	async load() {
		await this.content.load();
	}
	stop() {
		this.content.stop();
		for (const bundle of this.bundles) bundle.stop();
	}
	async flush(flushBundles = false) {
		await this.content.flush();
		for (const imp of this.imports) await imp.flush();
		await this.langBundle.flush();
		if (flushBundles) await Promise.all(this.bundles.map((bundle) => bundle.flush()));
	}
	switchActive(controller, resource) {
		this.activeController = controller;
		this.activeResource = resource;
		const pro = new Promise((resolve) => {
			this.once("switchActiveFinished", resolve);
		});
		this.emit("activeChanged");
		return pro;
	}
	allControllerNames(onlyWithAttaches = false) {
		return this.info.decls.filter((decl) => decl.type === "interface.controller").filter(onlyWithAttaches ? (decl) => decl.attachs.length > 0 : () => true).map((info) => info.name);
	}
	allResourceNames(checkController = "") {
		return this.info.decls.filter((decl) => decl.type === "interface.resource").filter(checkController ? (decl) => !decl.controller || decl.controller.includes(checkController) : () => true).map((info) => info.name);
	}
	updatePaths() {
		const ctrlInfo = this.info.decls.filter((decl) => decl.type === "interface.controller").find((info) => info.name === this.activeController);
		const resInfo = this.info.decls.filter((decl) => decl.type === "interface.resource").find((info) => info.name === this.activeResource);
		const finalPaths = [];
		if (resInfo) finalPaths.push(...resInfo.paths);
		if (ctrlInfo) finalPaths.push(...ctrlInfo.attachs);
		if (finalPaths.length > 0) {
			if (JSON.stringify(this.paths) === JSON.stringify(finalPaths)) {
				this.emit("switchActiveFinished");
				return;
			}
			for (const content of this.imports) content.stop();
			this.paths = finalPaths;
			this.bundles = this.paths.map((dir) => {
				return new Bundle(this.content.loader, this.content.watcher, this.maa, path.join(this.root, dir));
			});
		} else {
			for (const bundle of this.bundles) bundle.stop();
			this.paths = [];
			this.bundles = [];
		}
		this.emit("pathChanged");
	}
	updateLangs() {
		const newFiles = this.info.decls.filter((decl) => decl.type === "interface.language").map((info) => [info.name, info.path]);
		this.langBundle.update(newFiles).then(() => {
			this.emit("localeChanged");
		});
	}
	removeFile(file) {
		this.info.decls = this.info.decls.filter((decl) => decl.file !== file);
		this.info.refs = this.info.refs.filter((ref) => ref.file !== file);
		this.info.layer.removeFile(file);
	}
	updateImports() {
		const newFiles = this.info.refs.filter((ref) => ref.type === "interface.import_path").map((info) => info.target);
		if (JSON.stringify(this.importFiles) === JSON.stringify(newFiles)) return;
		this.info.decls = this.info.decls.filter((decl) => decl.file === this.file);
		this.info.refs = this.info.refs.filter((ref) => ref.file === this.file);
		for (const content of this.imports) {
			content.stop();
			this.info.layer.removeFile(content.file);
		}
		this.importFiles = newFiles;
		this.imports = newFiles.map((file) => {
			const full = joinPath(this.root, file);
			return new ContentJson(this.content.loader, this.content.watcher, full, (node) => {
				this.removeFile(full);
				if (node) parseInterface(node, this.info, {
					maa: this.maa,
					file: full,
					import: true
				});
				this.emit("slaveInterfaceChanged");
			});
		});
		this.emit("importChanged");
	}
	locateLayer(file) {
		const rel = relativePath(this.root, file).replaceAll(path.sep, "/");
		if (file === this.file || this.importFiles.includes(rel) || this.langBundle.langs.find((lang) => lang.file === rel)) return [
			this.info.layer,
			file,
			false
		];
		else for (const bundle of this.bundles) {
			if (file.startsWith(joinPath(bundle.root, this.maa ? "tasks" : "pipeline"))) return [
				bundle.layer,
				file,
				false
			];
			if (file === bundle.defaultPipelinePath) return [
				bundle.layer,
				file,
				true
			];
		}
		return null;
	}
	get allLayers() {
		const layers = this.bundles.map((bundle) => bundle.layer);
		layers.push(this.info.layer);
		return layers;
	}
	get topLayer() {
		return this.info.layer;
	}
	evalTask(task) {
		return this.topLayer.evalTask(task);
	}
	maaEvalTask(task) {
		if (!this.maa) return null;
		const result = new MaaEvalContext(this.eval).evalTask(task);
		if (result) delete result.task.__baseTaskResolved;
		return result;
	}
	maaEvalExpr(expr, self, strip) {
		if (!this.maa) return null;
		return new MaaEvalContext(this.eval).evalExpr(expr, self, strip);
	}
};

//#endregion
export { Bundle, BundleManager, FsContentLoader, FsContentWatcher, InterfaceBundle, LayerInfo, buildControllerRuntime, buildDiagnosticMessage, buildOption, buildResourceRuntime, buildTask, buildTaskRef, buildTaskRuntime, extractTaskRef, filterDeclRef, findDeclRef, findMaaDeclRef, isAnchorRef, isBool, isNumber, isString, joinImagePath, joinPath, normalizeImageFolder, parseArray, parseInterface, parseObject, parseObjectFlex, parseTask, performDiagnostic, relativePath, resolveCheckbox, resolveOptionConfig, resolveSelect };
